{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Shared setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains imports, constants and functions which are shared across all other notebooks containing analyses for both the main body of the paper and the supplementary information."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# python standard library\n",
    "import sys\n",
    "# ensure Python 3.4\n",
    "assert sys.version_info[0:2] == (3, 4)\n",
    "import os\n",
    "import operator\n",
    "import itertools\n",
    "import collections\n",
    "import functools\n",
    "import glob\n",
    "import csv\n",
    "import datetime\n",
    "import bisect\n",
    "from bisect import bisect_left, bisect_right\n",
    "import sqlite3\n",
    "import subprocess\n",
    "import random\n",
    "import gc\n",
    "import shutil\n",
    "import shelve\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# general purpose third party packages\n",
    "\n",
    "import cython\n",
    "assert cython.__version__ == '0.22'\n",
    "#%load_ext Cython\n",
    "\n",
    "import numpy as np\n",
    "assert np.__version__ == '1.9.2'\n",
    "nnz = np.count_nonzero\n",
    "\n",
    "import scipy\n",
    "assert scipy.__version__ == '0.15.1'\n",
    "import scipy.stats\n",
    "import scipy.spatial.distance\n",
    "from scipy import stats\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "import numexpr\n",
    "assert numexpr.__version__ == '2.4'\n",
    "\n",
    "import h5py\n",
    "assert h5py.__version__ == '2.4.0'\n",
    "\n",
    "import tables\n",
    "assert tables.__version__ == '3.1.1'\n",
    "\n",
    "import bcolz\n",
    "assert bcolz.__version__ == '0.8.1'\n",
    "\n",
    "import pandas\n",
    "assert pandas.__version__ == '0.16.0'\n",
    "\n",
    "import IPython\n",
    "assert IPython.__version__ == '3.0.0'\n",
    "# from IPython.html.widgets import interact, interactive\n",
    "# from IPython.html import widgets\n",
    "from IPython.display import clear_output, display, HTML\n",
    "\n",
    "import rpy2\n",
    "assert rpy2.__version__ == '2.5.4'\n",
    "import rpy2.robjects as ro\n",
    "%reload_ext rpy2.ipython\n",
    "\n",
    "import statsmodels\n",
    "assert statsmodels.__version__ == '0.6.1'\n",
    "\n",
    "import sklearn\n",
    "assert sklearn.__version__ == '0.15.2'\n",
    "import sklearn.decomposition\n",
    "import sklearn.manifold\n",
    "\n",
    "import sh\n",
    "assert sh.__version__ == '1.11'\n",
    "\n",
    "import sqlalchemy\n",
    "assert sqlalchemy.__version__ == '0.9.8'\n",
    "\n",
    "import pymysql\n",
    "assert pymysql.__version__ == '0.6.3.None'\n",
    "\n",
    "import psycopg2\n",
    "assert psycopg2.__version__ == '2.5.4 (dt dec pq3 ext)'\n",
    "\n",
    "import petl as etl\n",
    "assert etl.__version__ == '1.0.7'\n",
    "etl.config.display_index_header = True\n",
    "\n",
    "import humanize\n",
    "# VERSION (0, 4) doesn't match PyPI (0.5.1)\n",
    "from humanize import naturalsize, intcomma, intword"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plotting setup\n",
    "%matplotlib inline\n",
    "import matplotlib as mpl\n",
    "assert mpl.__version__ == '1.4.3'\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "assert sns.__version__ == '0.5.1'\n",
    "\n",
    "sns.set_context('paper')\n",
    "sns.set_style('white')\n",
    "sns.set_style('ticks')\n",
    "\n",
    "rcParams = plt.rcParams\n",
    "rcParams['font.size'] = 9\n",
    "rcParams['axes.labelsize'] = 9\n",
    "rcParams['xtick.labelsize'] = 9\n",
    "rcParams['ytick.labelsize'] = 9\n",
    "rcParams['legend.fontsize'] = 9\n",
    "rcParams['axes.linewidth'] = .5\n",
    "rcParams['lines.linewidth'] = .5\n",
    "rcParams['patch.linewidth'] = .5\n",
    "rcParams['font.family'] = 'arial'\n",
    "rcParams['ytick.direction'] = 'out'\n",
    "rcParams['xtick.direction'] = 'out'\n",
    "rcParams['savefig.jpeg_quality'] = 100\n",
    "rcParams['savefig.dpi'] = 120\n",
    "rcParams['lines.markeredgewidth'] = .5\n",
    "\n",
    "golden_ratio = 1.61803398875"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# bio third party packages\n",
    "\n",
    "import Bio\n",
    "assert Bio.__version__ == '1.65'\n",
    "\n",
    "import pyfasta\n",
    "# no version identifier, cannot verify version\n",
    "\n",
    "import pysam\n",
    "assert pysam.__version__ == '0.8.1'\n",
    "\n",
    "import pysamstats\n",
    "assert pysamstats.__version__ == '0.21'\n",
    "\n",
    "import petlx\n",
    "assert petlx.__version__ == '1.0.3'\n",
    "import petlx.bio\n",
    "\n",
    "import vcf\n",
    "assert vcf.VERSION == '0.6.7'\n",
    "\n",
    "import vcfnp\n",
    "assert vcfnp.__version__ == '2.1.2'\n",
    "\n",
    "import anhima\n",
    "assert anhima.__version__ == '0.11.1'\n",
    "\n",
    "import allel\n",
    "assert allel.__version__ == '0.12.0'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def log(*msg):\n",
    "    msg = str(datetime.datetime.now()) + ' :: ' + ' '.join(map(str, msg))\n",
    "    print(msg, file=sys.stderr) \n",
    "    sys.stderr.flush()\n",
    "    \n",
    "from IPython.display import display_html\n",
    "\n",
    "def pre(msg):\n",
    "    display_html('<pre style=\"line-height: 100%; font-size: .7em; display: inline-block; white-space: pre; background-color: #ff6; border: 0px solid #ddd; padding: 1px\">' + str(msg).replace('<', '&lt;') + '</pre>', raw=True)\n",
    "    \n",
    "def html(content):\n",
    "    display_html(content, raw=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "CROSSES = '3d7_hb3', 'hb3_dd2', '7g8_gb4'\n",
    "CHROMOSOMES = tuple('Pf3D7_{:02d}_v3'.format(n).encode('ascii') for n in range(1, 15))\n",
    "PARENT_RUNS = 'ERR019061', 'ERR019054', 'ERR012788', 'ERR012840', 'ERR027099', 'ERR027100'\n",
    "LABELS = {\n",
    "    '3d7_hb3': '3D7 x HB3', \n",
    "    'hb3_dd2': 'HB3 x Dd2', \n",
    "    '7g8_gb4': '7G8 x GB4',\n",
    "    'ERR019061': '3D7',\n",
    "    'ERR019054': 'HB3(1)',\n",
    "    'ERR012788': 'HB3(2)', \n",
    "    'ERR012840': 'Dd2',\n",
    "    'ERR027099': '7G8',\n",
    "    'ERR027100': 'GB4'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# N.B., change this if data are stored at a different location\n",
    "BASE_DIR = '/data/plasmodium/pfalciparum/pf-crosses'\n",
    "DATA_DIR = os.path.join(BASE_DIR, 'data')\n",
    "RELEASE_DIR = os.path.join(DATA_DIR, 'release')\n",
    "PUBLIC_DIR = os.path.join(DATA_DIR, 'public/20141022')\n",
    "\n",
    "# public data\n",
    "GATK_VCF_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.gatk.final.vcf.gz')\n",
    "GATK_CALLSET_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.gatk.final.npz')\n",
    "CORTEX_VCF_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.cortex.final.vcf.gz')\n",
    "CORTEX_CALLSET_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.cortex.final.npz')\n",
    "COMBINED_VCF_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.combined.final.vcf.gz')\n",
    "COMBINED_CALLSET_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.combined.final.npz')\n",
    "CORTEX_RAW_VCF_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.cortex.raw.vcf.gz')\n",
    "BAM_FN_TEMPLATE = os.path.join(PUBLIC_DIR, 'bam/{run}.realigned.bam')\n",
    "COVERAGE_FN_TEMPLATE = os.path.join(PUBLIC_DIR, 'cnv/{run}.coverage_ext_binned.300.{chrom}.npy')\n",
    "CNV_TBL_FN_TEMPLATE = os.path.join(PUBLIC_DIR, 'cnv/{run}.cnv.300.txt')\n",
    "CNV_FN_TEMPLATE = os.path.join(PUBLIC_DIR, 'cnv/{run}.cnv.300.{chrom}.npy')\n",
    "\n",
    "# reference data\n",
    "GENOME_DIR = os.path.join(DATA_DIR, 'genome/sanger/version3/September_2012')\n",
    "FASTA_FN = os.path.join(GENOME_DIR, 'Pf3D7_v3.fa')\n",
    "GFF_FN = os.path.join(GENOME_DIR, 'Pf3D7_v3.sorted.ascii.gff.gz')\n",
    "fasta = pyfasta.Fasta(FASTA_FN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dup_samples = {\n",
    "    '3d7_hb3': ['C01/PG0062-C/ERR019070',\n",
    "                'C02/PG0053-C/ERR019067',\n",
    "                'C02/PG0055-C/ERR019066',\n",
    "                'C02/PG0056-C/ERR019068'],\n",
    "    'hb3_dd2': [],\n",
    "    '7g8_gb4': ['AUD/PG0112-CW/ERR045639',\n",
    "                'JC9/PG0111-C/ERR029409',\n",
    "                'JE11/PG0100-CW/ERR045630',\n",
    "                'JF6/PG0079-CW/ERR045637',\n",
    "                'KB8/PG0104-CW/ERR045642',\n",
    "                'LA10/PG0086-CW/ERR045629',\n",
    "                'NIC/PG0095-C/ERR027107',\n",
    "                'QF5/PG0078-CW/ERR045638',\n",
    "                'XD8/PG0105-CW/ERR045628',\n",
    "                'XF12/PG0102-CW/ERR045635'],\n",
    "}\n",
    "\n",
    "\n",
    "excessive_recomb_samples = {\n",
    "    '3d7_hb3': [],\n",
    "    'hb3_dd2': ['SC01/PG0025-C/ERR019045'],\n",
    "    '7g8_gb4': ['D2/PG0094-CW/ERR045632'],\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Numpy utilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def interleave(a, b):\n",
    "    c = np.empty((a.size + b.size,), dtype=a.dtype)\n",
    "    c[0::2] = a\n",
    "    c[1::2] = b\n",
    "    return c\n",
    "\n",
    "\n",
    "def unpack_callset(callset):\n",
    "    V = callset['variants'].view(np.recarray)\n",
    "    C = callset['calldata'].view(np.recarray)\n",
    "    C2d = vcfnp.view2d(C).view(np.recarray)\n",
    "    return V, C, C2d\n",
    "\n",
    "\n",
    "def filter_variants(callset, query=None, verbose=True):\n",
    "    if query is None:\n",
    "        return callset\n",
    "    else:\n",
    "        V, C, _ = unpack_callset(callset)\n",
    "        if isinstance(query, str):\n",
    "            flt = numexpr.evaluate(query, local_dict=V)\n",
    "        elif callable(query):\n",
    "            flt = query(*unpack_callset(callset))\n",
    "        else:\n",
    "            flt = np.asarray(query)\n",
    "        indices = np.nonzero(flt)[0]\n",
    "        ret, exc, tot = indices.size, V.size-indices.size, V.size\n",
    "        if verbose:\n",
    "            log('filter variants: excluding %s (%.1f%%)' % (exc, exc*100./tot), 'retaining %s (%.1f%%)' % (ret, ret*100./tot), 'of', tot, 'variants')\n",
    "        V = np.take(V, indices, axis=0).view(np.recarray)\n",
    "        C = np.take(C, indices, axis=0).view(np.recarray)\n",
    "        return {'variants': V, 'calldata': C}\n",
    "\n",
    "\n",
    "def filter_calls(callset, query=None, verbose=True):\n",
    "    if query is None:\n",
    "        return callset\n",
    "    else:\n",
    "        V, C, _ = unpack_callset(callset)\n",
    "        C = C.copy()\n",
    "        C2d = vcfnp.view2d(C).view(np.recarray)\n",
    "\n",
    "        # filter calls\n",
    "        if isinstance(query, str):\n",
    "            flt = numexpr.evaluate(query, local_dict=C2d)\n",
    "        elif callable(query):\n",
    "            flt = query(*unpack_callset(callset))\n",
    "        else:\n",
    "            flt = np.asarray(query)\n",
    "        # only filter non-missing\n",
    "        excl = (C2d.genotype >= 0) & ~flt\n",
    "\n",
    "        tot, exc, ret = excl.size, np.count_nonzero(excl), np.count_nonzero(~excl)\n",
    "        if verbose:\n",
    "            log('filter calls: excluding %s (%.1f%%)' % (exc, exc*100./tot), 'retaining %s (%.1f%%)' % (ret, ret*100./tot), 'of', tot, 'calls')\n",
    "\n",
    "        C2d.genotype[excl] = -2\n",
    "        return {'variants': V, 'calldata': C}    \n",
    "    \n",
    "\n",
    "def filter_samples(callset, inclusions=None, exclusions=None, verbose=True):\n",
    "    if inclusions or exclusions:\n",
    "        V, C, _ = unpack_callset(callset)\n",
    "        samples = C.dtype.names\n",
    "        if inclusions:\n",
    "            exclusions = [s for s in samples if s not in inclusions]\n",
    "        elif exclusions:\n",
    "            inclusions = [s for s in samples if s not in exclusions]\n",
    "        if verbose:\n",
    "            log('filter samples: excluding', exclusions, 'including', inclusions)\n",
    "        C = C[inclusions].copy()\n",
    "        return {'variants': V, 'calldata': C}\n",
    "    else:\n",
    "        return callset\n",
    "    \n",
    "    \n",
    "def filter_callset(callset, \n",
    "                   variant_filter=None, \n",
    "                   call_filter=None, \n",
    "                   sample_inclusions=None, \n",
    "                   sample_exclusions=None,\n",
    "                   verbose=True):\n",
    "    \n",
    "    callset = filter_variants(callset, variant_filter, verbose=verbose)\n",
    "    \n",
    "    callset = filter_samples(callset, inclusions=sample_inclusions, exclusions=sample_exclusions, verbose=verbose)\n",
    "    \n",
    "    callset = filter_calls(callset, call_filter, verbose=verbose)\n",
    "    \n",
    "    return callset\n",
    "\n",
    "\n",
    "def load_callset(callset_fn, \n",
    "                 variant_filter=None, \n",
    "                 call_filter=None, \n",
    "                 sample_inclusions=None, \n",
    "                 sample_exclusions=None,\n",
    "                 verbose=True):\n",
    "    \n",
    "    if verbose:\n",
    "        log('loading', callset_fn)\n",
    "    callset = np.load(callset_fn)\n",
    "\n",
    "    callset = filter_callset(callset, \n",
    "                             variant_filter=variant_filter, \n",
    "                             call_filter=call_filter, \n",
    "                             sample_inclusions=sample_inclusions,\n",
    "                             sample_exclusions=sample_exclusions,\n",
    "                             verbose=verbose)\n",
    "    \n",
    "    return callset\n",
    "\n",
    "\n",
    "def filter_callsets(callsets,\n",
    "                    crosses=CROSSES,\n",
    "                    variant_filter=None, \n",
    "                    call_filter=None, \n",
    "                    sample_inclusions=dict(), \n",
    "                    sample_exclusions=dict(),\n",
    "                    verbose=True):\n",
    "    out = dict()\n",
    "    for cross in callsets.keys():\n",
    "        out[cross] = filter_callset(callsets[cross], \n",
    "                                    variant_filter=variant_filter,\n",
    "                                    call_filter=call_filter,\n",
    "                                    sample_inclusions=sample_inclusions.get(cross, None),\n",
    "                                    sample_exclusions=sample_exclusions.get(cross, None),\n",
    "                                    verbose=verbose)\n",
    "    return out\n",
    "    \n",
    "    \n",
    "def load_callsets(callset_fn_template,\n",
    "                  crosses=CROSSES,\n",
    "                  variant_filter=None, \n",
    "                  call_filter=None, \n",
    "                  sample_inclusions=dict(), \n",
    "                  sample_exclusions=dict(),\n",
    "                  verbose=True):\n",
    "    callsets = dict()\n",
    "    for cross in CROSSES:\n",
    "        callset_fn = callset_fn_template.format(cross=cross)\n",
    "        callsets[cross] = load_callset(callset_fn, \n",
    "                                       variant_filter=variant_filter,\n",
    "                                       call_filter=call_filter,\n",
    "                                       sample_inclusions=sample_inclusions.get(cross, None),\n",
    "                                       sample_exclusions=sample_exclusions.get(cross, None),\n",
    "                                       verbose=verbose)\n",
    "    return callsets\n",
    "    \n",
    "    \n",
    "def locate_variants(CHROM, POS, chrom, start=None, stop=None):\n",
    "    if isinstance(chrom, str):\n",
    "        chrom = chrom.encode('ascii')\n",
    "    start_index, stop_index = bisect_left(CHROM, chrom), bisect_right(CHROM, chrom)\n",
    "    POS = POS[start_index:stop_index]\n",
    "    if stop is not None:\n",
    "        stop_index = start_index + bisect_right(POS, stop)\n",
    "    if start is not None:\n",
    "        start_index = start_index + bisect_left(POS, start)\n",
    "    return start_index, stop_index\n",
    "\n",
    "\n",
    "def min_haplen_calls(min_haplen):\n",
    "    def _f(v, c, c2d):\n",
    "        hl, hs, _ = haplotypes(v, c)\n",
    "        return hl >= min_haplen\n",
    "    return _f\n",
    "\n",
    "\n",
    "def gatk_conf_calls(V, C, C2d):\n",
    "    return C2d.GQ >= 99\n",
    "\n",
    "\n",
    "def cortex_conf_calls(V, C, C2d):\n",
    "    return C2d.GT_CONF >= 50\n",
    "\n",
    "\n",
    "def combined_conf_calls(v, c, c2d):\n",
    "    return ((v.set[:,np.newaxis] == b'Cortex') & (c2d.GT_CONF >= 50)\n",
    "            | (v.set[:,np.newaxis] != b'Cortex') & (c2d.GQ >= 99))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def display_with_nrows(tbl, n=5, caption=''):\n",
    "    caption += ' (%s rows)' % tbl.nrows()\n",
    "    tbl.display(n, caption=caption)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "CACHE_DIR = os.path.join(PUBLIC_DIR, 'cache')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def etlcache(f):\n",
    "    fn = os.path.join(CACHE_DIR, f.__name__)\n",
    "    if not os.path.exists(fn):\n",
    "        etl.topickle(f(), fn)\n",
    "    return etl.frompickle(fn)\n",
    "    \n",
    "    \n",
    "def pdcache(f):\n",
    "    fn = os.path.join(CACHE_DIR, f.__name__)\n",
    "    if not os.path.exists(fn):\n",
    "        obj = f()\n",
    "        pandas.to_pickle(obj, fn)\n",
    "        return obj\n",
    "    else:\n",
    "        return pandas.read_pickle(fn)\n",
    "    \n",
    "    \n",
    "def nocache(f):\n",
    "    fn = os.path.join(CACHE_DIR, f.__name__)\n",
    "    if os.path.exists(fn):\n",
    "        os.remove(fn)\n",
    "    return f()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|cross</th>\n",
       "<th>1|clone</th>\n",
       "<th>2|sample</th>\n",
       "<th>3|run</th>\n",
       "<th>4|instrument</th>\n",
       "<th>5|coverage</th>\n",
       "<th>6|ID</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>3d7_hb3</td>\n",
       "<td>3D7</td>\n",
       "<td>PG0051-C</td>\n",
       "<td>ERR019061</td>\n",
       "<td>Illumina Genome Analyzer II</td>\n",
       "<td>122X</td>\n",
       "<td>3D7/PG0051-C/ERR019061</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>3d7_hb3</td>\n",
       "<td>C01</td>\n",
       "<td>PG0065-C</td>\n",
       "<td>ERR019064</td>\n",
       "<td>Illumina Genome Analyzer II</td>\n",
       "<td>163X</td>\n",
       "<td>C01/PG0065-C/ERR019064</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>3d7_hb3</td>\n",
       "<td>C01</td>\n",
       "<td>PG0062-C</td>\n",
       "<td>ERR019070</td>\n",
       "<td>Illumina Genome Analyzer II</td>\n",
       "<td>108X</td>\n",
       "<td>C01/PG0062-C/ERR019070</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>3d7_hb3</td>\n",
       "<td>C02</td>\n",
       "<td>PG0055-C</td>\n",
       "<td>ERR019066</td>\n",
       "<td>Illumina Genome Analyzer II</td>\n",
       "<td>102X</td>\n",
       "<td>C02/PG0055-C/ERR019066</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>3d7_hb3</td>\n",
       "<td>C02</td>\n",
       "<td>PG0053-C</td>\n",
       "<td>ERR019067</td>\n",
       "<td>Illumina Genome Analyzer II</td>\n",
       "<td>73X</td>\n",
       "<td>C02/PG0053-C/ERR019067</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+-----------+-------+------------+-------------+-------------------------------+----------+--------------------------+\n",
       "| cross     | clone | sample     | run         | instrument                    | coverage | ID                       |\n",
       "+===========+=======+============+=============+===============================+==========+==========================+\n",
       "| '3d7_hb3' | '3D7' | 'PG0051-C' | 'ERR019061' | 'Illumina Genome Analyzer II' | '122X'   | '3D7/PG0051-C/ERR019061' |\n",
       "+-----------+-------+------------+-------------+-------------------------------+----------+--------------------------+\n",
       "| '3d7_hb3' | 'C01' | 'PG0065-C' | 'ERR019064' | 'Illumina Genome Analyzer II' | '163X'   | 'C01/PG0065-C/ERR019064' |\n",
       "+-----------+-------+------------+-------------+-------------------------------+----------+--------------------------+\n",
       "| '3d7_hb3' | 'C01' | 'PG0062-C' | 'ERR019070' | 'Illumina Genome Analyzer II' | '108X'   | 'C01/PG0062-C/ERR019070' |\n",
       "+-----------+-------+------------+-------------+-------------------------------+----------+--------------------------+\n",
       "| '3d7_hb3' | 'C02' | 'PG0055-C' | 'ERR019066' | 'Illumina Genome Analyzer II' | '102X'   | 'C02/PG0055-C/ERR019066' |\n",
       "+-----------+-------+------------+-------------+-------------------------------+----------+--------------------------+\n",
       "| '3d7_hb3' | 'C02' | 'PG0053-C' | 'ERR019067' | 'Illumina Genome Analyzer II' | '73X'    | 'C02/PG0053-C/ERR019067' |\n",
       "+-----------+-------+------------+-------------+-------------------------------+----------+--------------------------+\n",
       "..."
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tbl_samples = (etl\n",
    "    .fromtsv(os.path.join(PUBLIC_DIR, 'samples.txt'))\n",
    "    .addfield('ID', lambda row: '%s/%s/%s' % (row.clone, row.sample, row.run))\n",
    ")\n",
    "run2id = tbl_samples.lookupone('run', 'ID')\n",
    "tbl_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genome region accessibility classifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|region_chrom</th>\n",
       "<th>1|region_start</th>\n",
       "<th>2|region_stop</th>\n",
       "<th>3|region_type</th>\n",
       "<th>4|region_size</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td style='text-align: right'>1</td>\n",
       "<td style='text-align: right'>27336</td>\n",
       "<td>SubtelomericRepeat</td>\n",
       "<td style='text-align: right'>27336</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td style='text-align: right'>27337</td>\n",
       "<td style='text-align: right'>92900</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td style='text-align: right'>92901</td>\n",
       "<td style='text-align: right'>457931</td>\n",
       "<td>Core</td>\n",
       "<td style='text-align: right'>365031</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td style='text-align: right'>457932</td>\n",
       "<td style='text-align: right'>460311</td>\n",
       "<td>Centromere</td>\n",
       "<td style='text-align: right'>2380</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td style='text-align: right'>460312</td>\n",
       "<td style='text-align: right'>575900</td>\n",
       "<td>Core</td>\n",
       "<td style='text-align: right'>115589</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+---------------+--------------+-------------+-----------------------------+-------------+\n",
       "| region_chrom  | region_start | region_stop | region_type                 | region_size |\n",
       "+===============+==============+=============+=============================+=============+\n",
       "| 'Pf3D7_01_v3' |            1 |       27336 | 'SubtelomericRepeat'        |       27336 |\n",
       "+---------------+--------------+-------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' |        27337 |       92900 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+--------------+-------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' |        92901 |      457931 | 'Core'                      |      365031 |\n",
       "+---------------+--------------+-------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' |       457932 |      460311 | 'Centromere'                |        2380 |\n",
       "+---------------+--------------+-------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' |       460312 |      575900 | 'Core'                      |      115589 |\n",
       "+---------------+--------------+-------------+-----------------------------+-------------+\n",
       "..."
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@etlcache\n",
    "def tbl_regions_1b():\n",
    "    return (etl\n",
    "        .fromtsv(os.path.join(PUBLIC_DIR, 'regions-20130225.onebased.txt'))\n",
    "        .pushheader(['chrom', 'start', 'stop', 'type'])\n",
    "        .convertnumbers()\n",
    "        .prefixheader('region_')\n",
    "        .addfield('region_size', lambda row: row.region_stop - row.region_start + 1)\n",
    "    )\n",
    "tbl_regions_1b"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genome annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|feature_chrom</th>\n",
       "<th>1|feature_type</th>\n",
       "<th>2|feature_start</th>\n",
       "<th>3|feature_stop</th>\n",
       "<th>4|feature_length</th>\n",
       "<th>5|feature_strand</th>\n",
       "<th>6|feature_id</th>\n",
       "<th>7|feature_parent_id</th>\n",
       "<th>8|feature_name</th>\n",
       "<th>9|feature_previous_id</th>\n",
       "<th>10|feature_region_type</th>\n",
       "<th>11|feature_region_size</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>repeat_region</td>\n",
       "<td style='text-align: right'>1</td>\n",
       "<td style='text-align: right'>360</td>\n",
       "<td style='text-align: right'>359</td>\n",
       "<td>+</td>\n",
       "<td>Pfalciparum_REP_20</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericRepeat</td>\n",
       "<td style='text-align: right'>27336</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>repeat_region</td>\n",
       "<td style='text-align: right'>361</td>\n",
       "<td style='text-align: right'>1418</td>\n",
       "<td style='text-align: right'>1057</td>\n",
       "<td>+</td>\n",
       "<td>Pfalciparum_REP_15</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericRepeat</td>\n",
       "<td style='text-align: right'>27336</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>repeat_region</td>\n",
       "<td style='text-align: right'>2160</td>\n",
       "<td style='text-align: right'>3858</td>\n",
       "<td style='text-align: right'>1698</td>\n",
       "<td>+</td>\n",
       "<td>Pfalciparum_REP_35</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericRepeat</td>\n",
       "<td style='text-align: right'>27336</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>repeat_region</td>\n",
       "<td style='text-align: right'>8856</td>\n",
       "<td style='text-align: right'>9021</td>\n",
       "<td style='text-align: right'>165</td>\n",
       "<td>+</td>\n",
       "<td>Pfalciparum_REP_5</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericRepeat</td>\n",
       "<td style='text-align: right'>27336</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>repeat_region</td>\n",
       "<td style='text-align: right'>9313</td>\n",
       "<td style='text-align: right'>9529</td>\n",
       "<td style='text-align: right'>216</td>\n",
       "<td>+</td>\n",
       "<td>Pfalciparum_REP_25</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericRepeat</td>\n",
       "<td style='text-align: right'>27336</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+---------------+-----------------+---------------+--------------+----------------+----------------+----------------------+-------------------+--------------+---------------------+----------------------+---------------------+\n",
       "| feature_chrom | feature_type    | feature_start | feature_stop | feature_length | feature_strand | feature_id           | feature_parent_id | feature_name | feature_previous_id | feature_region_type  | feature_region_size |\n",
       "+===============+=================+===============+==============+================+================+======================+===================+==============+=====================+======================+=====================+\n",
       "| 'Pf3D7_01_v3' | 'repeat_region' |             1 |          360 |            359 | '+'            | 'Pfalciparum_REP_20' | None              | None         | None                | 'SubtelomericRepeat' |               27336 |\n",
       "+---------------+-----------------+---------------+--------------+----------------+----------------+----------------------+-------------------+--------------+---------------------+----------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'repeat_region' |           361 |         1418 |           1057 | '+'            | 'Pfalciparum_REP_15' | None              | None         | None                | 'SubtelomericRepeat' |               27336 |\n",
       "+---------------+-----------------+---------------+--------------+----------------+----------------+----------------------+-------------------+--------------+---------------------+----------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'repeat_region' |          2160 |         3858 |           1698 | '+'            | 'Pfalciparum_REP_35' | None              | None         | None                | 'SubtelomericRepeat' |               27336 |\n",
       "+---------------+-----------------+---------------+--------------+----------------+----------------+----------------------+-------------------+--------------+---------------------+----------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'repeat_region' |          8856 |         9021 |            165 | '+'            | 'Pfalciparum_REP_5'  | None              | None         | None                | 'SubtelomericRepeat' |               27336 |\n",
       "+---------------+-----------------+---------------+--------------+----------------+----------------+----------------------+-------------------+--------------+---------------------+----------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'repeat_region' |          9313 |         9529 |            216 | '+'            | 'Pfalciparum_REP_25' | None              | None         | None                | 'SubtelomericRepeat' |               27336 |\n",
       "+---------------+-----------------+---------------+--------------+----------------+----------------+----------------------+-------------------+--------------+---------------------+----------------------+---------------------+\n",
       "..."
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@etlcache\n",
    "def tbl_features():\n",
    "    return (etl\n",
    "        .fromgff3(GFF_FN)\n",
    "        .unpackdict('attributes', ['ID', 'Parent', 'Name', 'previous_systematic_id'])\n",
    "        .rename({'seqid': 'chrom', 'end': 'stop', 'ID': 'id', 'Parent': 'parent_id', 'Name': 'name', 'previous_systematic_id': 'previous_id'})\n",
    "        .cutout('source', 'score', 'phase')\n",
    "        .addfield('length', lambda row: row.stop - row.start, index=4)\n",
    "        .intervalleftjoin(tbl_regions_1b, lkey='chrom', rkey='region_chrom', rstart='region_start', rstop='region_stop')\n",
    "        .cutout('region_chrom', 'region_start', 'region_stop')\n",
    "        .distinct('id')\n",
    "        .sort(key=('chrom', 'start', 'type'))\n",
    "        .prefixheader('feature_')\n",
    "    )\n",
    "tbl_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "lkp_feature = tbl_features.recordlookupone('feature_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|feature_chrom</th>\n",
       "<th>1|feature_type</th>\n",
       "<th>2|feature_start</th>\n",
       "<th>3|feature_stop</th>\n",
       "<th>4|feature_length</th>\n",
       "<th>5|feature_strand</th>\n",
       "<th>6|feature_id</th>\n",
       "<th>7|feature_name</th>\n",
       "<th>8|feature_previous_id</th>\n",
       "<th>9|feature_region_type</th>\n",
       "<th>10|feature_region_size</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>gene</td>\n",
       "<td style='text-align: right'>29510</td>\n",
       "<td style='text-align: right'>37126</td>\n",
       "<td style='text-align: right'>7616</td>\n",
       "<td>+</td>\n",
       "<td>PF3D7_0100100</td>\n",
       "<td>VAR</td>\n",
       "<td>PFA0005w</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>gene</td>\n",
       "<td style='text-align: right'>38982</td>\n",
       "<td style='text-align: right'>40207</td>\n",
       "<td style='text-align: right'>1225</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100200</td>\n",
       "<td>RIF</td>\n",
       "<td>PFA0010c</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>gene</td>\n",
       "<td style='text-align: right'>42367</td>\n",
       "<td style='text-align: right'>46507</td>\n",
       "<td style='text-align: right'>4140</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100300</td>\n",
       "<td>None</td>\n",
       "<td>PFA0015c</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>gene</td>\n",
       "<td style='text-align: right'>50363</td>\n",
       "<td style='text-align: right'>51636</td>\n",
       "<td style='text-align: right'>1273</td>\n",
       "<td>+</td>\n",
       "<td>PF3D7_0100400</td>\n",
       "<td>RIF</td>\n",
       "<td>PFA0020w</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>pseudogene</td>\n",
       "<td style='text-align: right'>53169</td>\n",
       "<td style='text-align: right'>53280</td>\n",
       "<td style='text-align: right'>111</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100500</td>\n",
       "<td>var pseudogene, exon 1</td>\n",
       "<td>PFA0025c</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+---------------+--------------+---------------+--------------+----------------+----------------+-----------------+--------------------------+---------------------+-----------------------------+---------------------+\n",
       "| feature_chrom | feature_type | feature_start | feature_stop | feature_length | feature_strand | feature_id      | feature_name             | feature_previous_id | feature_region_type         | feature_region_size |\n",
       "+===============+==============+===============+==============+================+================+=================+==========================+=====================+=============================+=====================+\n",
       "| 'Pf3D7_01_v3' | 'gene'       |         29510 |        37126 |           7616 | '+'            | 'PF3D7_0100100' | 'VAR'                    | 'PFA0005w'          | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-----------------+--------------------------+---------------------+-----------------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'gene'       |         38982 |        40207 |           1225 | '-'            | 'PF3D7_0100200' | 'RIF'                    | 'PFA0010c'          | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-----------------+--------------------------+---------------------+-----------------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'gene'       |         42367 |        46507 |           4140 | '-'            | 'PF3D7_0100300' | None                     | 'PFA0015c'          | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-----------------+--------------------------+---------------------+-----------------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'gene'       |         50363 |        51636 |           1273 | '+'            | 'PF3D7_0100400' | 'RIF'                    | 'PFA0020w'          | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-----------------+--------------------------+---------------------+-----------------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'pseudogene' |         53169 |        53280 |            111 | '-'            | 'PF3D7_0100500' | 'var pseudogene, exon 1' | 'PFA0025c'          | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-----------------+--------------------------+---------------------+-----------------------------+---------------------+\n",
       "..."
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@etlcache\n",
    "def tbl_genes():\n",
    "    return (\n",
    "        tbl_features\n",
    "        .selectin('feature_type', ['gene', 'pseudogene'])\n",
    "        .cutout('feature_parent_id')\n",
    "    )\n",
    "tbl_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|feature_chrom</th>\n",
       "<th>1|feature_type</th>\n",
       "<th>2|feature_start</th>\n",
       "<th>3|feature_stop</th>\n",
       "<th>4|feature_length</th>\n",
       "<th>5|feature_id</th>\n",
       "<th>6|feature_region_size</th>\n",
       "<th>7|feature_region_type</th>\n",
       "<th>8|region_size</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intergenic_0</td>\n",
       "<td style='text-align: right'>37126</td>\n",
       "<td style='text-align: right'>38982</td>\n",
       "<td style='text-align: right'>1856</td>\n",
       "<td>PF3D7_0100100_PF3D7_0100200</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intergenic_1</td>\n",
       "<td style='text-align: right'>40207</td>\n",
       "<td style='text-align: right'>42367</td>\n",
       "<td style='text-align: right'>2160</td>\n",
       "<td>PF3D7_0100200_PF3D7_0100300</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intergenic_2</td>\n",
       "<td style='text-align: right'>46507</td>\n",
       "<td style='text-align: right'>50363</td>\n",
       "<td style='text-align: right'>3856</td>\n",
       "<td>PF3D7_0100300_PF3D7_0100400</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intergenic_0</td>\n",
       "<td style='text-align: right'>51636</td>\n",
       "<td style='text-align: right'>53169</td>\n",
       "<td style='text-align: right'>1533</td>\n",
       "<td>PF3D7_0100400_PF3D7_0100500</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intergenic_1</td>\n",
       "<td style='text-align: right'>53280</td>\n",
       "<td style='text-align: right'>53778</td>\n",
       "<td style='text-align: right'>498</td>\n",
       "<td>PF3D7_0100500_PF3D7_0100600</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+---------------+----------------+---------------+--------------+----------------+-------------------------------+---------------------+-----------------------------+-------------+\n",
       "| feature_chrom | feature_type   | feature_start | feature_stop | feature_length | feature_id                    | feature_region_size | feature_region_type         | region_size |\n",
       "+===============+================+===============+==============+================+===============================+=====================+=============================+=============+\n",
       "| 'Pf3D7_01_v3' | 'intergenic_0' |         37126 |        38982 |           1856 | 'PF3D7_0100100_PF3D7_0100200' |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+----------------+---------------+--------------+----------------+-------------------------------+---------------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' | 'intergenic_1' |         40207 |        42367 |           2160 | 'PF3D7_0100200_PF3D7_0100300' |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+----------------+---------------+--------------+----------------+-------------------------------+---------------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' | 'intergenic_2' |         46507 |        50363 |           3856 | 'PF3D7_0100300_PF3D7_0100400' |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+----------------+---------------+--------------+----------------+-------------------------------+---------------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' | 'intergenic_0' |         51636 |        53169 |           1533 | 'PF3D7_0100400_PF3D7_0100500' |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+----------------+---------------+--------------+----------------+-------------------------------+---------------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' | 'intergenic_1' |         53280 |        53778 |            498 | 'PF3D7_0100500_PF3D7_0100600' |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+----------------+---------------+--------------+----------------+-------------------------------+---------------------+-----------------------------+-------------+\n",
       "..."
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@etlcache\n",
    "def tbl_intergenic():\n",
    "    return (\n",
    "        tbl_genes\n",
    "        .addfieldusingcontext('next', lambda prv, cur, nxt: (nxt.feature_start, nxt.feature_id, nxt.feature_strand) if nxt and cur.feature_chrom == nxt.feature_chrom else None)\n",
    "        .notnone('next')\n",
    "        .unpack('next', ['next_start', 'next_id', 'next_strand'])\n",
    "        .convert('feature_id', lambda v, row: '%s_%s' % (row.feature_id, row.next_id), pass_row=True)\n",
    "        .cutout('feature_start', 'feature_length', 'feature_region_type', 'next_id')\n",
    "        .rename({'feature_stop': 'feature_start', 'next_start': 'feature_stop'})\n",
    "        .movefield('feature_stop', 3)\n",
    "        .addfield('feature_length', lambda row: row.feature_stop - row.feature_start, index=4)\n",
    "        .convert('feature_type', lambda v, row: 'intergenic %s/%s' % (row.feature_strand, row.next_strand), pass_row=True)\n",
    "        # reclassify based on expected number of promoters\n",
    "        .convert('feature_type', {'intergenic +/-': 'intergenic_0', 'intergenic -/+': 'intergenic_2', 'intergenic +/+': 'intergenic_1', 'intergenic -/-': 'intergenic_1'})\n",
    "        .cutout('feature_strand', 'next_strand', 'feature_name', 'feature_previous_id')\n",
    "        .intervalleftjoin(tbl_regions_1b, \n",
    "                          lkey='feature_chrom', lstart='feature_start', lstop='feature_stop',\n",
    "                          rkey='region_chrom', rstart='region_start', rstop='region_stop')\n",
    "        .rename('region_type', 'feature_region_type')\n",
    "        .cutout('region_chrom', 'region_start', 'region_stop')\n",
    "        .gt('feature_length', 0)\n",
    "    )\n",
    "\n",
    "tbl_intergenic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|feature_chrom</th>\n",
       "<th>1|feature_type</th>\n",
       "<th>2|feature_start</th>\n",
       "<th>3|feature_stop</th>\n",
       "<th>4|feature_length</th>\n",
       "<th>5|feature_strand</th>\n",
       "<th>6|feature_id</th>\n",
       "<th>7|feature_parent_id</th>\n",
       "<th>8|feature_name</th>\n",
       "<th>9|feature_previous_id</th>\n",
       "<th>10|feature_region_type</th>\n",
       "<th>11|feature_region_size</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>29510</td>\n",
       "<td style='text-align: right'>34762</td>\n",
       "<td style='text-align: right'>5252</td>\n",
       "<td>+</td>\n",
       "<td>PF3D7_0100100.1:exon:1</td>\n",
       "<td>PF3D7_0100100.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>35888</td>\n",
       "<td style='text-align: right'>37126</td>\n",
       "<td style='text-align: right'>1238</td>\n",
       "<td>+</td>\n",
       "<td>PF3D7_0100100.1:exon:2</td>\n",
       "<td>PF3D7_0100100.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>38982</td>\n",
       "<td style='text-align: right'>39923</td>\n",
       "<td style='text-align: right'>941</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100200.1:exon:1</td>\n",
       "<td>PF3D7_0100200.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>40154</td>\n",
       "<td style='text-align: right'>40207</td>\n",
       "<td style='text-align: right'>53</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100200.1:exon:2</td>\n",
       "<td>PF3D7_0100200.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>42367</td>\n",
       "<td style='text-align: right'>43617</td>\n",
       "<td style='text-align: right'>1250</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100300.1:exon:1</td>\n",
       "<td>PF3D7_0100300.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+---------------+--------------+---------------+--------------+----------------+----------------+--------------------------+-------------------+--------------+---------------------+-----------------------------+---------------------+\n",
       "| feature_chrom | feature_type | feature_start | feature_stop | feature_length | feature_strand | feature_id               | feature_parent_id | feature_name | feature_previous_id | feature_region_type         | feature_region_size |\n",
       "+===============+==============+===============+==============+================+================+==========================+===================+==============+=====================+=============================+=====================+\n",
       "| 'Pf3D7_01_v3' | 'CDS'        |         29510 |        34762 |           5252 | '+'            | 'PF3D7_0100100.1:exon:1' | 'PF3D7_0100100.1' | None         | None                | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+--------------------------+-------------------+--------------+---------------------+-----------------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'CDS'        |         35888 |        37126 |           1238 | '+'            | 'PF3D7_0100100.1:exon:2' | 'PF3D7_0100100.1' | None         | None                | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+--------------------------+-------------------+--------------+---------------------+-----------------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'CDS'        |         38982 |        39923 |            941 | '-'            | 'PF3D7_0100200.1:exon:1' | 'PF3D7_0100200.1' | None         | None                | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+--------------------------+-------------------+--------------+---------------------+-----------------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'CDS'        |         40154 |        40207 |             53 | '-'            | 'PF3D7_0100200.1:exon:2' | 'PF3D7_0100200.1' | None         | None                | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+--------------------------+-------------------+--------------+---------------------+-----------------------------+---------------------+\n",
       "| 'Pf3D7_01_v3' | 'CDS'        |         42367 |        43617 |           1250 | '-'            | 'PF3D7_0100300.1:exon:1' | 'PF3D7_0100300.1' | None         | None                | 'SubtelomericHypervariable' |               65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+--------------------------+-------------------+--------------+---------------------+-----------------------------+---------------------+\n",
       "..."
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@etlcache\n",
    "def tbl_exons():\n",
    "    return (\n",
    "        tbl_features\n",
    "        .eq('feature_type', 'CDS')\n",
    "        .gt('feature_length', 0)\n",
    "    )\n",
    "\n",
    "tbl_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|feature_chrom</th>\n",
       "<th>1|feature_type</th>\n",
       "<th>2|feature_start</th>\n",
       "<th>3|feature_stop</th>\n",
       "<th>4|feature_length</th>\n",
       "<th>5|feature_strand</th>\n",
       "<th>6|feature_id</th>\n",
       "<th>7|feature_parent_id</th>\n",
       "<th>8|feature_name</th>\n",
       "<th>9|feature_previous_id</th>\n",
       "<th>10|feature_region_size</th>\n",
       "<th>11|feature_region_type</th>\n",
       "<th>12|region_size</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>34762</td>\n",
       "<td style='text-align: right'>35888</td>\n",
       "<td style='text-align: right'>1126</td>\n",
       "<td>+</td>\n",
       "<td>PF3D7_0100100.1:exon:1_PF3D7_0100100.1:exon:2</td>\n",
       "<td>PF3D7_0100100.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>39923</td>\n",
       "<td style='text-align: right'>40154</td>\n",
       "<td style='text-align: right'>231</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100200.1:exon:1_PF3D7_0100200.1:exon:2</td>\n",
       "<td>PF3D7_0100200.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>43617</td>\n",
       "<td style='text-align: right'>43775</td>\n",
       "<td style='text-align: right'>158</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100300.1:exon:1_PF3D7_0100300.1:exon:2</td>\n",
       "<td>PF3D7_0100300.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>50416</td>\n",
       "<td style='text-align: right'>50572</td>\n",
       "<td style='text-align: right'>156</td>\n",
       "<td>+</td>\n",
       "<td>PF3D7_0100400.1:exon:1_PF3D7_0100400.1:exon:2</td>\n",
       "<td>PF3D7_0100400.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>Pf3D7_01_v3</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>54788</td>\n",
       "<td style='text-align: right'>54938</td>\n",
       "<td style='text-align: right'>150</td>\n",
       "<td>-</td>\n",
       "<td>PF3D7_0100600.1:exon:1_PF3D7_0100600.1:exon:2</td>\n",
       "<td>PF3D7_0100600.1</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "<td>SubtelomericHypervariable</td>\n",
       "<td style='text-align: right'>65564</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+---------------+--------------+---------------+--------------+----------------+----------------+-------------------------------------------------+-------------------+--------------+---------------------+---------------------+-----------------------------+-------------+\n",
       "| feature_chrom | feature_type | feature_start | feature_stop | feature_length | feature_strand | feature_id                                      | feature_parent_id | feature_name | feature_previous_id | feature_region_size | feature_region_type         | region_size |\n",
       "+===============+==============+===============+==============+================+================+=================================================+===================+==============+=====================+=====================+=============================+=============+\n",
       "| 'Pf3D7_01_v3' | 'intron'     |         34762 |        35888 |           1126 | '+'            | 'PF3D7_0100100.1:exon:1_PF3D7_0100100.1:exon:2' | 'PF3D7_0100100.1' | None         | None                |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-------------------------------------------------+-------------------+--------------+---------------------+---------------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' | 'intron'     |         39923 |        40154 |            231 | '-'            | 'PF3D7_0100200.1:exon:1_PF3D7_0100200.1:exon:2' | 'PF3D7_0100200.1' | None         | None                |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-------------------------------------------------+-------------------+--------------+---------------------+---------------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' | 'intron'     |         43617 |        43775 |            158 | '-'            | 'PF3D7_0100300.1:exon:1_PF3D7_0100300.1:exon:2' | 'PF3D7_0100300.1' | None         | None                |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-------------------------------------------------+-------------------+--------------+---------------------+---------------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' | 'intron'     |         50416 |        50572 |            156 | '+'            | 'PF3D7_0100400.1:exon:1_PF3D7_0100400.1:exon:2' | 'PF3D7_0100400.1' | None         | None                |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-------------------------------------------------+-------------------+--------------+---------------------+---------------------+-----------------------------+-------------+\n",
       "| 'Pf3D7_01_v3' | 'intron'     |         54788 |        54938 |            150 | '-'            | 'PF3D7_0100600.1:exon:1_PF3D7_0100600.1:exon:2' | 'PF3D7_0100600.1' | None         | None                |               65564 | 'SubtelomericHypervariable' |       65564 |\n",
       "+---------------+--------------+---------------+--------------+----------------+----------------+-------------------------------------------------+-------------------+--------------+---------------------+---------------------+-----------------------------+-------------+\n",
       "..."
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@etlcache\n",
    "def tbl_introns():\n",
    "    return (\n",
    "        tbl_exons\n",
    "        .addfieldusingcontext('next', \n",
    "                              lambda prv, cur, nxt: (nxt.feature_start, nxt.feature_id) if nxt and cur.feature_parent_id == nxt.feature_parent_id else None)\n",
    "        .notnone('next')\n",
    "        .unpack('next', ['next_start', 'next_id'])\n",
    "        .convert('feature_id', lambda v, row: '%s_%s' % (row.feature_id, row.next_id), pass_row=True)\n",
    "        .cutout('feature_start', 'feature_length', 'feature_region_type', 'next_id')\n",
    "        .rename({'feature_stop': 'feature_start', 'next_start': 'feature_stop'})\n",
    "        .movefield('feature_stop', 3)\n",
    "        .addfield('feature_length', lambda row: row.feature_stop - row.feature_start, index=4)\n",
    "        .convert('feature_type', lambda v: 'intron')\n",
    "        .intervalleftjoin(tbl_regions_1b, \n",
    "                          lkey='feature_chrom', lstart='feature_start', lstop='feature_stop',\n",
    "                          rkey='region_chrom', rstart='region_start', rstop='region_stop')\n",
    "        .rename('region_type', 'feature_region_type')\n",
    "        .cutout('region_chrom', 'region_start', 'region_stop')\n",
    "    )\n",
    "\n",
    "tbl_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|feature_id</th>\n",
       "<th>1|feature_alias</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>PF3D7_1314600</td>\n",
       "<td>814061</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PF3D7_1314600</td>\n",
       "<td>PF13_0083</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PF3D7_0209400</td>\n",
       "<td>8444970</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PF3D7_0209400</td>\n",
       "<td>PF02_0090</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PF3D7_0209400</td>\n",
       "<td>PFB0423c</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+-----------------+---------------+\n",
       "| feature_id      | feature_alias |\n",
       "+=================+===============+\n",
       "| 'PF3D7_1314600' | '814061'      |\n",
       "+-----------------+---------------+\n",
       "| 'PF3D7_1314600' | 'PF13_0083'   |\n",
       "+-----------------+---------------+\n",
       "| 'PF3D7_0209400' | '8444970'     |\n",
       "+-----------------+---------------+\n",
       "| 'PF3D7_0209400' | 'PF02_0090'   |\n",
       "+-----------------+---------------+\n",
       "| 'PF3D7_0209400' | 'PFB0423c'    |\n",
       "+-----------------+---------------+\n",
       "..."
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@etlcache\n",
    "def tbl_gene_aliases():\n",
    "    return (etl\n",
    "        .fromtsv(os.path.join(DATA_DIR, 'genome/plasmodb/release-9.3/Pfalciparum3D7/txt/PlasmoDB-9.3_Pfalciparum3D7_GeneAliases.txt'))\n",
    "        .pushheader(['feature_id'] + list(map(str, range(10))))\n",
    "        .convert('feature_id', lambda v: v[:13])  # ignore alternate splice forms\n",
    "        .cat()  # square up short rows\n",
    "        .melt('feature_id')\n",
    "        .rename('value', 'feature_alias')\n",
    "        .cut('feature_id', 'feature_alias')\n",
    "        .ne('feature_alias', None)\n",
    "    )\n",
    "\n",
    "tbl_gene_aliases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Heterochromatin (Flueck et al. 2009)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|feature_alias</th>\n",
       "<th>1|hp1</th>\n",
       "<th>2|feature_id</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>PFA0005w</td>\n",
       "<td style='text-align: right'>4.648</td>\n",
       "<td>PF3D7_0100100</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PFA0010c</td>\n",
       "<td style='text-align: right'>3.763</td>\n",
       "<td>PF3D7_0100200</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PFA0015c</td>\n",
       "<td style='text-align: right'>5.042</td>\n",
       "<td>PF3D7_0100300</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PFA0020w</td>\n",
       "<td style='text-align: right'>3.721</td>\n",
       "<td>PF3D7_0100400</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PFA0030c</td>\n",
       "<td style='text-align: right'>4.398</td>\n",
       "<td>PF3D7_0100600</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+---------------+-------+-----------------+\n",
       "| feature_alias | hp1   | feature_id      |\n",
       "+===============+=======+=================+\n",
       "| 'PFA0005w'    | 4.648 | 'PF3D7_0100100' |\n",
       "+---------------+-------+-----------------+\n",
       "| 'PFA0010c'    | 3.763 | 'PF3D7_0100200' |\n",
       "+---------------+-------+-----------------+\n",
       "| 'PFA0015c'    | 5.042 | 'PF3D7_0100300' |\n",
       "+---------------+-------+-----------------+\n",
       "| 'PFA0020w'    | 3.721 | 'PF3D7_0100400' |\n",
       "+---------------+-------+-----------------+\n",
       "| 'PFA0030c'    | 4.398 | 'PF3D7_0100600' |\n",
       "+---------------+-------+-----------------+\n",
       "..."
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@etlcache\n",
    "def tbl_hp1():\n",
    "    return (etl\n",
    "        .fromtext(os.path.join(DATA_DIR, 'reference/flueck_2009/journal.ppat.1000569.s008.txt'))\n",
    "        .rowslice(3, None)\n",
    "        .ne('lines', '')\n",
    "        .capture('lines', r'^\\S+\\s+\\d+\\s+\\d+\\s+(\\S+)', ['feature_alias'], include_original=1)\n",
    "        .capture('lines', r'^\\S+\\s+\\d+\\s+\\d+\\s+\\S+.*?(-?\\d\\.\\d{3})', ['hp1'], include_original=1)\n",
    "        .convert('hp1', float)\n",
    "        .cutout('lines')\n",
    "        .leftjoin(tbl_gene_aliases, key='feature_alias')\n",
    "        .ne('feature_id', None)\n",
    "        .sort('feature_id')\n",
    "    )\n",
    "\n",
    "tbl_hp1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def plot_variants_locator(ax, callset, chrom, start=0, stop=None, step=5):\n",
    "    V, _, _ = unpack_callset(callset)\n",
    "    if stop is None:\n",
    "        stop = len(fasta[chrom])\n",
    "    indices = np.nonzero((V.CHROM == chrom.encode('ascii')) & (V.POS >= start) & (V.POS <= stop))[0]\n",
    "    POS = np.take(V.POS, indices, axis=0)\n",
    "    anhima.loc.plot_variant_locator(POS, ax=ax, step=step, start_position=start, stop_position=stop, line_args=dict(color='k'))\n",
    "    ax.set_xticks([])\n",
    "    ax.set_xlabel('')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "INHERITANCE_COLORS = [\n",
    "    'white', # missing\n",
    "    'grey', # filtered\n",
    "    'yellow', # parent missing\n",
    "    'yellow', # parent filtered\n",
    "    'black', # non-parental\n",
    "    'orange', # non-segregating ref \n",
    "    'green', # non-segregating alt \n",
    "    'blue', # parent 1 \n",
    "    'red' # parent 2 \n",
    "]\n",
    "\n",
    "\n",
    "def plot_inheritance(ax, callset, chrom, start=0, stop=None, sample_indices=None, colors=INHERITANCE_COLORS, alpha=1, **kwargs):\n",
    "\n",
    "    V, C, C2d = unpack_callset(callset)\n",
    "    samples = C.dtype.names\n",
    "    \n",
    "    if stop is None:\n",
    "        stop = len(fasta[chrom])\n",
    "    indices = np.nonzero((V.CHROM == chrom.encode('ascii')) & (V.POS >= start) & (V.POS <= stop))[0]\n",
    "    POS = np.take(V.POS, indices, axis=0)\n",
    "    G = np.take(C2d.genotype, indices, axis=0)\n",
    "    P = inheritance(G)\n",
    "    \n",
    "    if sample_indices is not None:\n",
    "        P = np.take(P, sample_indices, axis=1)\n",
    "        samples = np.take(samples, sample_indices, axis=0) \n",
    "\n",
    "    anhima.gt.plot_discrete_calldata(P, labels=samples, colors=colors, states=range(len(colors)), ax=ax, **kwargs)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_features(chrom, start=None, stop=None):\n",
    "    if None not in {start, stop}:\n",
    "        region = '%s:%s-%s' % (chrom, start, stop)\n",
    "    else:\n",
    "        region = chrom\n",
    "    tbl_features = (etl\n",
    "        .fromgff3(GFF_FN, region)\n",
    "        .cutout('source', 'score', 'phase')\n",
    "        .unpackdict('attributes', ['ID', 'Name', 'previous_systematic_id'], includeoriginal=True)\n",
    "    )\n",
    "    tbl_features = tbl_features.setheader(['feature_' + str(f).lower() for f in tbl_features.header()])\n",
    "    return tbl_features\n",
    "\n",
    "\n",
    "prog_hv = re.compile('var|rif|stev', flags=re.I) \n",
    "\n",
    "\n",
    "def plot_genes(ax, chrom, start=0, stop=None, label_named=False, exclude_labels=[], width=.3, label_ytextpad=.1,\n",
    "               normal_color='k', hypervariable_color='r', cds_color='gray', divider_linewidth=1,\n",
    "               annotate_genes=[], annotate_xtext=0, annotate_ytext_fwd=6, annotate_ytext_rev=8, \n",
    "               annotate_arrowprops=None, annotate_bbox=None, \n",
    "               show_cds=False, gene_labels=dict()):\n",
    "    if stop is None:\n",
    "        stop = len(fasta[chrom])\n",
    "    tbl_local_features = get_features(chrom, start, stop)\n",
    "    tbl_local_genes = tbl_local_features.selectin('feature_type', {'gene', 'pseudogene'})\n",
    "    tbl_local_cds = tbl_local_features.eq('feature_type', 'CDS')\n",
    "    \n",
    "    # divider\n",
    "    ax.plot([start, stop], [.5, .5], 'k-', linewidth=divider_linewidth)\n",
    "    ax.set_ylim(0, 1)\n",
    "    \n",
    "    # positive strand\n",
    "    tbl_fwd_genes = tbl_local_genes.eq('feature_strand', '+')\n",
    "    xranges = [(r.feature_start, r.feature_end-r.feature_start) for r in tbl_fwd_genes.records()]\n",
    "    fc = [hypervariable_color if r.feature_name is not None and prog_hv.search(r.feature_name) is not None else normal_color \n",
    "          for r in tbl_fwd_genes.records()]\n",
    "    ax.broken_barh(xranges, (.5, width), color=fc)\n",
    "    if show_cds:\n",
    "        tbl_fwd_cds = tbl_local_cds.eq('feature_strand', '+')\n",
    "        xranges = [(r.feature_start, r.feature_end-r.feature_start) for r in tbl_fwd_cds.records()]\n",
    "        ax.broken_barh(xranges, (.5, width), color=cds_color)\n",
    "        \n",
    "    \n",
    "    # negative strand\n",
    "    tbl_rev_genes = tbl_local_genes.eq('feature_strand', '-')\n",
    "    xranges = [(r.feature_start, r.feature_end-r.feature_start) for r in tbl_rev_genes.records()]\n",
    "    fc = [hypervariable_color if r.feature_name is not None and prog_hv.search(r.feature_name) is not None else normal_color\n",
    "          for r in tbl_rev_genes.records()]\n",
    "    ax.broken_barh(xranges, (.5-width, width), color=fc)\n",
    "    if show_cds:\n",
    "        tbl_rev_cds = tbl_local_cds.eq('feature_strand', '-')\n",
    "        xranges = [(r.feature_start, r.feature_end-r.feature_start) for r in tbl_rev_cds.records()]\n",
    "        ax.broken_barh(xranges, (.5-width, width), color=cds_color)\n",
    "    \n",
    "    for s in 'top', 'left', 'right':\n",
    "        ax.spines[s].set_visible(False)\n",
    "    ax.set_xlim(start, stop)\n",
    "    ax.set_yticks([])\n",
    "    ax.xaxis.tick_bottom()\n",
    "\n",
    "    # label named genes\n",
    "    if label_named:\n",
    "        for g in tbl_local_genes.records():\n",
    "            if g.feature_name is not None and g.feature_name not in exclude_labels:\n",
    "                x = g.feature_start\n",
    "                y = .5 + width + label_ytextpad if g.feature_strand == '+' else .5 - width - label_ytextpad\n",
    "                va = 'bottom' if g.feature_strand == '+' else 'top'\n",
    "                l = g.feature_name\n",
    "                ax.text(x, y, l, rotation=0, ha='left', va=va, transform=ax.transData)\n",
    "                \n",
    "    # annotate specific genes\n",
    "    if annotate_arrowprops is None:\n",
    "        annotate_arrowprops = dict(facecolor='k', arrowstyle='-', shrinkA=0, shrinkB=0)\n",
    "    if annotate_bbox is None:\n",
    "        annotate_bbox = dict(facecolor='w', alpha=.9)\n",
    "    for id in annotate_genes:\n",
    "        g = lkp_feature[id]\n",
    "        gene_id = g['feature_id']\n",
    "        l = gene_labels[gene_id] if gene_id in gene_labels else g['feature_name'] if g['feature_name'] is not None else gene_id\n",
    "        if g['feature_strand'] == '+':\n",
    "            ax.annotate(l, xy=(g['feature_start'], .5), ha='left', xytext=(annotate_xtext, annotate_ytext_fwd), \n",
    "                        xycoords='data', textcoords='offset points', \n",
    "                        arrowprops=annotate_arrowprops,\n",
    "                        bbox=annotate_bbox)\n",
    "        elif g['feature_strand'] == '-':\n",
    "            ax.annotate(l, xy=(g['feature_start'], .5), ha='left', xytext=(annotate_xtext, annotate_ytext_rev), \n",
    "                        xycoords='data', textcoords='offset points', \n",
    "                        arrowprops=annotate_arrowprops,\n",
    "                        bbox=annotate_bbox)\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Accessibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "CNV_RUNS = {\n",
    "    '3d7_hb3': ['ERR019061', 'ERR019054'],\n",
    "    'hb3_dd2': ['ERR012788', 'ERR012840', 'ERR019054'],  # share HB3s\n",
    "    '7g8_gb4': ['ERR027099', 'ERR027100']\n",
    "}\n",
    "\n",
    "\n",
    "accessibility_colors = {\n",
    "    'Core': 'white', \n",
    "    'SubtelomericRepeat': 'yellow',\n",
    "    'SubtelomericHypervariable': '#d73027',\n",
    "    'InternalHypervariable': '#fc8d59',\n",
    "    'Centromere': 'black',\n",
    "    'CNV': '#91bfdb',\n",
    "}\n",
    "\n",
    "\n",
    "def plot_accessibility(ax, chrom, cross=None, start=0, stop=None, linewidth=0.5, **kwargs):\n",
    "    tbl_lreg = tbl_regions_1b.eq('region_chrom', chrom)\n",
    "    if stop is None:\n",
    "        stop = len(fasta[chrom])\n",
    "\n",
    "    # plot core\n",
    "    xranges = tbl_lreg.eq('region_type', 'Core').values(['region_start', 'region_size'])\n",
    "    ax.broken_barh(xranges, (.1, .8), facecolor=accessibility_colors['Core'], \n",
    "                   linewidth=linewidth, **kwargs)\n",
    "\n",
    "    # plot centromere\n",
    "    xranges = tbl_lreg.eq('region_type', 'Centromere').values(['region_start', 'region_size'])\n",
    "    ax.broken_barh(xranges, (.1, .8), facecolor=accessibility_colors['Centromere'], \n",
    "                   linewidth=linewidth, **kwargs)\n",
    "\n",
    "    # plot subtelomeric repeat\n",
    "    xranges = tbl_lreg.eq('region_type', 'SubtelomericRepeat').values(['region_start', 'region_size'])\n",
    "    ax.broken_barh(xranges, (.1, .8), facecolor=accessibility_colors['SubtelomericRepeat'], \n",
    "                   linewidth=linewidth, **kwargs)\n",
    "    \n",
    "    # plot hypervariable\n",
    "    xranges = tbl_lreg.eq('region_type', 'SubtelomericHypervariable').values(['region_start', 'region_size'])\n",
    "    ax.broken_barh(xranges, (.1, .8), facecolor=accessibility_colors['SubtelomericHypervariable'], \n",
    "                   linewidth=linewidth, **kwargs)\n",
    "    xranges = tbl_lreg.eq('region_type', 'InternalHypervariable').values(['region_start', 'region_size'])\n",
    "    ax.broken_barh(xranges, (.1, .8), facecolor=accessibility_colors['InternalHypervariable'], \n",
    "                   linewidth=linewidth, **kwargs)\n",
    "    \n",
    "    if cross is not None:\n",
    "        # plot CNVs\n",
    "        tbls_cnvs = [etl.fromtsv(CNV_TBL_FN_TEMPLATE.format(run=run)).convertnumbers() for run in CNV_RUNS[cross]]\n",
    "        tbl_cnvs = (etl\n",
    "            .cat(*tbls_cnvs)\n",
    "            .gt('cn', 1)\n",
    "            .eq('chrom', chrom)\n",
    "            .intervalleftjoin(tbl_lreg, lkey='chrom', rkey='region_chrom', \n",
    "                              lstart='start', lstop='stop', \n",
    "                              rstart='region_start', rstop='region_stop')\n",
    "            .eq('region_type', 'Core')\n",
    "        )\n",
    "        xranges = tbl_cnvs.values(['start', 'size'])\n",
    "        ax.broken_barh(xranges, (.1, .8), color=accessibility_colors['CNV'], \n",
    "                       linewidth=linewidth, **kwargs)\n",
    "    \n",
    "    # tidy up\n",
    "    for s in 'top', 'left', 'right', 'bottom':\n",
    "        ax.spines[s].set_visible(False)\n",
    "    ax.set_xlim(start, stop)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.xaxis.tick_bottom()\n",
    "    ax.set_ylim(0, 1)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Heterochromatin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def plot_hp1(ax, chrom, start=0, stop=None):\n",
    "\n",
    "    if stop is None:\n",
    "        stop = len(fasta[chrom])\n",
    "        \n",
    "    tbl_genes_local = get_features(chrom, start, stop).selectin('feature_type', {'gene', 'pseudogene'})\n",
    "    A = tbl_genes_local.cut('feature_start', 'feature_end', 'feature_id').join(tbl_hp1, key='feature_id').torecarray()\n",
    "    left = A.feature_start\n",
    "    height = A.hp1\n",
    "    width = A.feature_end - A.feature_start\n",
    "    ax.bar(left, height, width, facecolor='purple', linewidth=1, edgecolor='purple')\n",
    "    \n",
    "    ax.set_xlim(start, stop)\n",
    "    for s in 'top', 'left', 'right', 'bottom':\n",
    "        ax.spines[s].set_visible(False)\n",
    "    ax.set_yticks([])\n",
    "    ax.set_ylim(-1, 7)\n",
    "    ax.set_xticks([])\n",
    "#    ax.plot([start, stop], [0, 0], color='k', linestyle='-')\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Inheritance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "INHERITANCE_COLORS = [\n",
    "    'white', # missing\n",
    "    'grey', # filtered\n",
    "    'yellow', # parent missing\n",
    "    'yellow', # parent filtered\n",
    "    'black', # non-parental\n",
    "    'orange', # non-segregating ref \n",
    "    'green', # non-segregating alt \n",
    "    'blue', # parent 1 \n",
    "    'red' # parent 2 \n",
    "]\n",
    "\n",
    "\n",
    "def plot_inheritance(ax, callset, chrom, start=0, stop=None, sample_indices=None, colors=INHERITANCE_COLORS, alpha=1, **kwargs):\n",
    "\n",
    "    V, C, C2d = unpack_callset(callset)\n",
    "    samples = C.dtype.names\n",
    "    \n",
    "    if stop is None:\n",
    "        stop = len(fasta[chrom])\n",
    "    indices = np.nonzero((V.CHROM == chrom) & (V.POS >= start) & (V.POS <= stop))[0]\n",
    "    POS = np.take(V.POS, indices, axis=0)\n",
    "    G = np.take(C2d.genotype, indices, axis=0)\n",
    "    P = inheritance(G)\n",
    "    \n",
    "    if sample_indices is not None:\n",
    "        P = np.take(P, sample_indices, axis=1)\n",
    "        samples = np.take(samples, sample_indices, axis=0) \n",
    "\n",
    "    vcfplt.discrete_calldata_colormesh(P, labels=samples, colors=colors, states=range(len(colors)), alpha=alpha, ax=ax, **kwargs)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recombination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def marker_span(callset, chrom):\n",
    "    V, _, _ = unpack_callset(callset)\n",
    "    start_index, stop_index = locate_variants(V.CHROM, V.POS, chrom)\n",
    "    POS = V.POS[start_index:stop_index]\n",
    "    x = POS[-1] - POS[0]\n",
    "    return x\n",
    "\n",
    "\n",
    "def marker_offset(callset, chrom):\n",
    "    V, _, _ = unpack_callset(callset)\n",
    "    start_index, stop_index = locate_variants(V.CHROM, V.POS, chrom)\n",
    "    POS = V.POS[start_index:stop_index]\n",
    "    return POS[0]\n",
    "\n",
    "\n",
    "def locate_flanking_markers(CHROM, POS, chrom, pos):\n",
    "    chrom_start_index, chrom_stop_index = bisect_left(CHROM, chrom), bisect_right(CHROM, chrom)\n",
    "    POS = POS[chrom_start_index:chrom_stop_index]\n",
    "    \n",
    "    start_index = bisect_left(POS, pos) - 1\n",
    "    stop_index = bisect_right(POS, pos)\n",
    "    \n",
    "    if start_index < 0:\n",
    "        start_index = None\n",
    "    else:\n",
    "        start_index += chrom_start_index\n",
    "        \n",
    "    if stop_index >= len(POS):\n",
    "        stop_index = None\n",
    "    else:\n",
    "        stop_index += chrom_start_index\n",
    "    \n",
    "    return start_index, stop_index\n",
    "    \n",
    "\n",
    "def simulate_random_uniform_crossovers(n_events, callset, chromosomes=CHROMOSOMES):\n",
    "    \n",
    "    # table of results\n",
    "    tbl = [['chrom', 'co_pos', 'co_pos_mid', 'co_pos_min', 'co_pos_max', 'co_pos_range']]\n",
    "    \n",
    "    # utility variables\n",
    "    V = callset['variants']\n",
    "    marker_spans = [marker_span(callset, chrom) for chrom in chromosomes]\n",
    "    marker_offsets = [marker_offset(callset, chrom) for chrom in chromosomes]\n",
    "    cumulative_marker_span = [np.sum(marker_spans[:i+1]) for i in range(len(marker_spans))]\n",
    "    total_marker_span = np.sum(marker_spans)\n",
    "    \n",
    "    # generate random integers over the total marker span\n",
    "    X = np.random.randint(low=0, high=total_marker_span, size=n_events)\n",
    "    \n",
    "    # map each event onto a chromosome and position\n",
    "    for x in X:\n",
    "\n",
    "        # which chromosome?\n",
    "        chrom_index = bisect_left(cumulative_marker_span, x)\n",
    "        chrom = chromosomes[chrom_index]\n",
    "\n",
    "        # which position?\n",
    "        if chrom_index > 0:\n",
    "            pos_relative = x - cumulative_marker_span[chrom_index-1]\n",
    "        else:\n",
    "            pos_relative = x\n",
    "        pos = pos_relative + marker_offsets[chrom_index]\n",
    "        \n",
    "        # find flanking markers\n",
    "        lidx, ridx = locate_flanking_markers(V.CHROM, V.POS, chrom, pos)\n",
    "\n",
    "        pos_min = V.POS[lidx]\n",
    "        pos_max = V.POS[ridx]\n",
    "        pos_mid = (pos_min + pos_max)//2\n",
    "        pos_range = pos_max - pos_min\n",
    "        \n",
    "        row = [chrom, pos, pos_mid, pos_min, pos_max, pos_range]\n",
    "        tbl += [row]\n",
    "        \n",
    "    return etl.wrap(tbl)\n",
    "\n",
    "\n",
    "def simulate_random_uniform_non_crossovers(n_events, callset, phi, chromosomes=CHROMOSOMES):\n",
    "    \n",
    "    # table of results\n",
    "    tbl = [['chrom', \n",
    "            'tract_start', \n",
    "            'tract_stop', \n",
    "            'tract_length',\n",
    "            'tract_start_min', \n",
    "            'tract_start_mid', \n",
    "            'tract_start_max', \n",
    "            'tract_stop_min', \n",
    "            'tract_stop_mid', \n",
    "            'tract_stop_max', \n",
    "            'tract_length_min', \n",
    "            'tract_length_mid', \n",
    "            'tract_length_max', \n",
    "            'tract_support']]\n",
    "    \n",
    "    # utility variables\n",
    "    V = callset['variants']\n",
    "    marker_spans = [marker_span(callset, chrom) for chrom in chromosomes]\n",
    "    marker_offsets = [marker_offset(callset, chrom) for chrom in chromosomes]\n",
    "    cumulative_marker_span = [np.sum(marker_spans[:i+1]) for i in range(len(marker_spans))]\n",
    "    total_marker_span = np.sum(marker_spans)\n",
    "    \n",
    "    # generate random integers over the total marker span\n",
    "    X = np.random.randint(low=0, high=total_marker_span, size=n_events)\n",
    "\n",
    "    # generate random geometric tract lengths\n",
    "    lengths = stats.geom.rvs(1-phi, size=n_events).astype('int32')\n",
    "    \n",
    "    # map each event onto a chromosome and position\n",
    "    for x, length in zip(X, lengths):\n",
    "\n",
    "        try:\n",
    "            # which chromosome?\n",
    "            chrom_index = bisect_left(cumulative_marker_span, x)\n",
    "            chrom = chromosomes[chrom_index]\n",
    "\n",
    "            # which start position?\n",
    "            if chrom_index > 0:\n",
    "                pos_relative = x - cumulative_marker_span[chrom_index-1]\n",
    "            else:\n",
    "                pos_relative = x\n",
    "            start = pos_relative + marker_offsets[chrom_index]\n",
    "            stop = start + length\n",
    "\n",
    "            # find flanking markers\n",
    "            start_lidx, start_ridx = locate_flanking_markers(V.CHROM, V.POS, chrom, start)\n",
    "            stop_lidx, stop_ridx = locate_flanking_markers(V.CHROM, V.POS, chrom, stop)\n",
    "            \n",
    "            if None in {start_lidx, start_ridx, stop_lidx, stop_ridx}:\n",
    "                \n",
    "                # falls off edge of marker span\n",
    "                support = start_min = start_max = start_mid = stop_min = stop_max = stop_mid = length_min = length_max = length_mid = None\n",
    "\n",
    "            else:\n",
    "\n",
    "                support = stop_ridx - start_ridx\n",
    "\n",
    "                # do we even observe the event?\n",
    "                if support == 0:\n",
    "                    \n",
    "                    start_min = start_max = start_mid = stop_min = stop_max = stop_mid = length_min = length_max = length_mid = None\n",
    "\n",
    "                else:\n",
    "\n",
    "                    start_min = V.POS[start_lidx]\n",
    "                    start_max = V.POS[start_ridx]\n",
    "                    start_mid = (start_min + start_max)//2\n",
    "                    stop_min = V.POS[stop_lidx]\n",
    "                    stop_max = V.POS[stop_ridx]\n",
    "                    stop_mid = (stop_min + stop_max)//2\n",
    "                    length_min = stop_min - start_max\n",
    "                    length_max = stop_max - start_min\n",
    "                    length_mid = (length_min + length_max)//2\n",
    "\n",
    "            row = [chrom, \n",
    "                   start, \n",
    "                   stop, \n",
    "                   length,\n",
    "                   start_min, \n",
    "                   start_mid, \n",
    "                   start_max, \n",
    "                   stop_min, \n",
    "                   stop_mid, \n",
    "                   stop_max, \n",
    "                   length_min, \n",
    "                   length_mid, \n",
    "                   length_max, \n",
    "                   support]\n",
    "            tbl += [row]\n",
    "    \n",
    "        except Exception as e:\n",
    "            log(e)\n",
    "            log(traceback.format_exc(10))\n",
    "                    \n",
    "    return etl.wrap(tbl)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optimisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Cython magic has been move to the Cython package, hence \n",
      "`%load_ext cythonmagic` is deprecated; Please use `%load_ext Cython` instead.\n",
      "\n",
      "Though, because I am nice, I'll still try to load it for you this time.\n"
     ]
    }
   ],
   "source": [
    "%reload_ext cythonmagic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%cython\n",
    "\n",
    "import numpy as np\n",
    "cimport numpy as np\n",
    "import petl as etl\n",
    "import vcfnp\n",
    "\n",
    "\n",
    "CHROMOSOMES = [('Pf3D7_%02d_v3' % n).encode('ascii') for n in range(1, 15)]\n",
    "\n",
    "# constants to represent the possible outcomes when attempting to determine the inheritance for any given variant call\n",
    "INHERITANCE_MISSING = 0  # genotype call is missing\n",
    "INHERITANCE_FILTERED = 1  # genotype call is filtered\n",
    "INHERITANCE_PARENT_MISSING = 2  # genotype is called but cannot be phased because one or both parents is missing\n",
    "INHERITANCE_PARENT_FILTERED = 3  # genotype is called but cannot be phased because one or both parents is filtered\n",
    "INHERITANCE_NONPARENTAL = 4  # genotype call does not correspond with either parent (i.e., mendelian error sensu strictu)\n",
    "INHERITANCE_NONSEGREGATING_REF = 5  # genotype call is ref and parents both have reference allele\n",
    "INHERITANCE_NONSEGREGATING_ALT = 6  # genotype call is alt and parents both have alternate allele\n",
    "INHERITANCE_PARENT1 = 7  # genotype call corresponds with first parent\n",
    "INHERITANCE_PARENT2 = 8  # genotype call corresponds with second parent\n",
    "\n",
    "\n",
    "# functions to determine inheritance states\n",
    "###########################################\n",
    "\n",
    "\n",
    "def inheritance(np.ndarray[np.int8_t, ndim=2] G):\n",
    "    cdef int m = G.shape[0] # number of variants\n",
    "    cdef int n = G.shape[1] # number of samples\n",
    "    cdef np.ndarray[np.uint8_t, ndim=2] P\n",
    "    P = np.zeros((m,n), dtype='u1')\n",
    "    for i in range(m):\n",
    "        for j in range(n):\n",
    "            # N.B., parents are *first* two samples\n",
    "            P[i,j] = inheritance_state(G[i,j], G[i,0], G[i,1])\n",
    "    return P\n",
    "\n",
    "\n",
    "cdef int inheritance_state(int genotype_child, int genotype_parent1, int genotype_parent2):\n",
    "    if genotype_child == -1:\n",
    "        return INHERITANCE_MISSING\n",
    "    elif genotype_child == -2:\n",
    "        return INHERITANCE_FILTERED\n",
    "    elif genotype_parent1 == -1 or genotype_parent2 == -1:\n",
    "        return INHERITANCE_PARENT_MISSING\n",
    "    elif genotype_parent1 == -2 or genotype_parent2 == -2:\n",
    "        return INHERITANCE_PARENT_FILTERED\n",
    "    elif genotype_child != genotype_parent1 and genotype_child != genotype_parent2:\n",
    "        return INHERITANCE_NONPARENTAL\n",
    "    elif genotype_child == genotype_parent1 == genotype_parent2 == 0:\n",
    "        return INHERITANCE_NONSEGREGATING_REF\n",
    "    elif genotype_child == genotype_parent1 == genotype_parent2 and genotype_child > 0:\n",
    "        return INHERITANCE_NONSEGREGATING_ALT\n",
    "    elif genotype_child == genotype_parent1:\n",
    "        return INHERITANCE_PARENT1\n",
    "    elif genotype_child == genotype_parent2:\n",
    "        return INHERITANCE_PARENT2\n",
    "    else:\n",
    "        return -1 # should never be reached\n",
    "\n",
    "\n",
    "# functions to calculate haplotype statistics\n",
    "#############################################\n",
    "\n",
    "\n",
    "def haplotypes_chrom(V, C, chrom):\n",
    "    \"\"\"\n",
    "    Construct arrays of haplotype length and haplotype support for all variant calls.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    tbl = list()\n",
    "    header = 'sample', 'chrom', 'start_min', 'start_mid', 'start_max', 'stop_min', 'stop_mid', 'stop_max', 'length_min', \\\n",
    "             'length_mid', 'length_max', 'support', 'prv_inheritance', 'inheritance', 'nxt_inheritance'\n",
    "    tbl.append(header)\n",
    "\n",
    "    cdef int i, j, M, N, prv_idx, start_idx, stop_idx, start_min, start_mid, start_max, stop_min, stop_mid, \\\n",
    "        stop_max, length_min, length_mid, length_max, support, prv, cur\n",
    "\n",
    "    samples = C.dtype.names\n",
    "    C2d = vcfnp.view2d(C).view(np.recarray)\n",
    "\n",
    "    # only deal with one chromosome, make life simpler\n",
    "    flt = V.CHROM == chrom\n",
    "    indices = np.nonzero(flt)[0]\n",
    "    V = np.take(V, indices, axis=0)\n",
    "    C2d = np.take(C2d, indices, axis=0)\n",
    "    M, N = C2d.shape\n",
    "    cdef np.ndarray[np.uint8_t, ndim=2] P\n",
    "    P = inheritance(C2d.genotype)  # inheritance\n",
    "#    print M, N\n",
    "\n",
    "    cdef np.ndarray[np.int32_t, ndim=1] POS\n",
    "    POS = V['POS']\n",
    "\n",
    "    cdef np.ndarray[np.int32_t, ndim=2] HL\n",
    "    cdef np.ndarray[np.int32_t, ndim=2] HS\n",
    "    HL = np.zeros((M, N), dtype='i4') # haplotype lengths (minimum)\n",
    "    HS = np.zeros((M, N), dtype='i4') # haplotype support - number of calls supporting the haplotype\n",
    "\n",
    "    cdef np.ndarray[np.int32_t, ndim=1] previous_call_idx\n",
    "    cdef np.ndarray[np.int32_t, ndim=2] haplotype_start_idxs\n",
    "    cdef np.ndarray[np.int32_t, ndim=1] last_haplotype_inheritance\n",
    "    previous_call_idx = np.array([-1] * N, dtype='i4') # store index of last good call\n",
    "    haplotype_start_idxs = np.array([[-1, -1]] * N, dtype='i4') # store indices of haplotype start (flanking markers)\n",
    "    last_haplotype_inheritance = np.array([-1] * N, dtype='i4') # store parental inheritance of upstream block\n",
    "\n",
    "    # iterate over variants\n",
    "    for i in range(M):\n",
    "        # iterate over samples\n",
    "        for j in range(N):\n",
    "            # extract genotype call for current variant/sample\n",
    "            cur = P[i, j]\n",
    "            if cur == INHERITANCE_PARENT1 or cur == INHERITANCE_PARENT2: # current call is good\n",
    "                # index of last good call for this sample\n",
    "                prv_idx = previous_call_idx[j]\n",
    "                if prv_idx < 0:\n",
    "                    # this is the first good call, start a haplotype\n",
    "                    haplotype_start_idxs[j] = (i, i)\n",
    "                else:\n",
    "                    # extract last good genotype call for current sample\n",
    "                    prv = P[prv_idx, j]\n",
    "                    if cur != prv:\n",
    "                        # inheritance switch, haplotype has ended\n",
    "                        start_min_idx, start_max_idx = tuple(haplotype_start_idxs[j])\n",
    "                        stop_min_idx, stop_max_idx = prv_idx, i\n",
    "                        start_min = POS[start_min_idx]\n",
    "                        start_max = POS[start_max_idx]\n",
    "                        start_mid = (start_max + start_min) // 2\n",
    "                        stop_min = POS[stop_min_idx]\n",
    "                        stop_max = POS[stop_max_idx]\n",
    "                        stop_mid = (stop_max + stop_min) // 2\n",
    "                        length_min = stop_min - start_max\n",
    "                        length_max = stop_max - start_min\n",
    "                        length_mid = stop_mid - start_mid\n",
    "                        support = np.count_nonzero(P[start_max_idx:stop_min_idx+1, j] == prv)\n",
    "                        # assign results\n",
    "                        HL[start_max_idx:stop_min_idx+1, j] = length_min\n",
    "                        HS[start_max_idx:stop_min_idx+1, j] = support\n",
    "                        row = samples[j], chrom, start_min, start_mid, start_max, stop_min, stop_mid, stop_max, \\\n",
    "                              length_min, length_mid, length_max, support, last_haplotype_inheritance[j], prv, cur\n",
    "                        tbl.append(row)\n",
    "                        # start new haplotype\n",
    "                        haplotype_start_idxs[j] = (prv_idx, i)\n",
    "                        last_haplotype_inheritance[j] = prv\n",
    "                    else:\n",
    "                        # haplotype continues\n",
    "                        pass\n",
    "                # store last good call index\n",
    "                previous_call_idx[j] = i\n",
    "\n",
    "    # final haplotypes ending at chromosome end\n",
    "    for j in range(N):\n",
    "        start_min_idx, start_max_idx = tuple(haplotype_start_idxs[j])\n",
    "        stop_min_idx, stop_max_idx = prv_idx, prv_idx  # nothing subsequent\n",
    "        start_min = POS[start_min_idx]\n",
    "        start_max = POS[start_max_idx]\n",
    "        start_mid = (start_max + start_min) // 2\n",
    "        stop_min = POS[stop_min_idx]\n",
    "        stop_max = POS[stop_max_idx]\n",
    "        stop_mid = (stop_max + stop_min) // 2\n",
    "        length_min = stop_min - start_max\n",
    "        length_max = stop_max - start_min\n",
    "        length_mid = stop_mid - start_mid\n",
    "        inherit = P[prv_idx, j]\n",
    "        support = np.count_nonzero(P[start_max_idx:stop_min_idx+1, j] == inherit)\n",
    "        # assign results\n",
    "        HL[start_max_idx:stop_min_idx+1, j] = length_min\n",
    "        HS[start_max_idx:stop_min_idx+1, j] = support\n",
    "        row = samples[j], chrom, start_min, start_mid, start_max, stop_min, stop_mid, stop_max, length_min, length_mid, \\\n",
    "              length_max, support, last_haplotype_inheritance[j], inherit, -1\n",
    "        tbl.append(row)\n",
    "\n",
    "    return HL, HS, etl.wrap(tbl)\n",
    "\n",
    "\n",
    "def haplotypes(V, C):\n",
    "    stats = [haplotype_stats_chrom(V, C, chrom) for chrom in CHROMOSOMES]\n",
    "    HL = np.vstack(s[0] for s in stats)\n",
    "    HS = np.vstack(s[1] for s in stats)\n",
    "    tbl = etl.cat(*[s[2] for s in stats])\n",
    "    return HL, HS, tbl\n",
    "\n",
    "\n",
    "# backwards compatibility\n",
    "haplotype_stats_chrom = haplotypes_chrom\n",
    "haplotype_stats = haplotypes\n",
    "\n",
    "\n",
    "# functions to tabulate recombination events\n",
    "############################################\n",
    "\n",
    "\n",
    "def tabulate_switches_chrom(V, C, chrom):\n",
    "    tbl = list()\n",
    "    header = 'sample', 'chrom', 'pos', 'lpos', 'rpos', 'range', 'from', 'to'\n",
    "    tbl.append(header)\n",
    "\n",
    "    cdef int i, j, m, n, prv_idx, cur, prv\n",
    "\n",
    "    samples = C.dtype.names\n",
    "    C2d = vcfnp.view2d(C).view(np.recarray)\n",
    "\n",
    "    # only deal with one chromosome, make life simpler\n",
    "    flt = V.CHROM == chrom\n",
    "    indices = np.nonzero(flt)[0]\n",
    "    V, C, C2d = [np.take(X, indices, axis=0) for X in [V, C, C2d]]\n",
    "    m, n = C2d.shape\n",
    "\n",
    "    cdef np.ndarray[np.uint8_t, ndim=2] P\n",
    "    P = inheritance(C2d.genotype)\n",
    "\n",
    "    cdef np.ndarray[np.int32_t, ndim=1] POS\n",
    "    POS = V['POS']\n",
    "\n",
    "    cdef np.ndarray[np.int32_t, ndim=1] previous_call_idx\n",
    "    previous_call_idx = np.array([-1] * n, dtype='i4') # store index of last good call\n",
    "\n",
    "    # iterate over variants\n",
    "    for i in range(m):\n",
    "        # iterate over samples\n",
    "        for j in range(n):\n",
    "            # extract genotype call for current variant/sample\n",
    "            cur = P[i, j]\n",
    "            if cur == INHERITANCE_PARENT1 or cur == INHERITANCE_PARENT2: # current call is good\n",
    "                # index of last good call for this sample\n",
    "                prv_idx = previous_call_idx[j]\n",
    "                if prv_idx < 0:\n",
    "                    # this is the first good call\n",
    "                    pass\n",
    "                else:\n",
    "                    # extract last good genotype call for current sample\n",
    "                    prv = P[prv_idx, j]\n",
    "                    if cur != prv:\n",
    "                        # phase switch\n",
    "                        lpos = POS[prv_idx]\n",
    "                        rpos = POS[i]\n",
    "                        row = samples[j], chrom, (lpos + rpos) // 2, lpos, rpos, rpos-lpos, prv, cur\n",
    "                        tbl.append(row)\n",
    "                    else:\n",
    "                        # haplotype continues\n",
    "                        pass\n",
    "                # store last good call index\n",
    "                previous_call_idx[j] = i\n",
    "\n",
    "    return etl.wrap(tbl)\n",
    "\n",
    "\n",
    "def tabulate_switches(V, C):\n",
    "    tbls = [tabulate_switches_chrom(V, C, chrom) for chrom in CHROMOSOMES]\n",
    "    return etl.cat(*tbls)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
