{
 "metadata": {
  "name": "",
  "signature": "sha256:7d830ee36bb0b5be9c91b0f595be590bc8a427755e7d3408ae7ab7a30c7bec4c"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Imports"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from __future__ import division\n",
      "import sys\n",
      "import os\n",
      "import glob\n",
      "import csv\n",
      "import re\n",
      "import shelve\n",
      "import traceback\n",
      "from datetime import datetime\n",
      "import math\n",
      "import numpy as np\n",
      "import numexpr\n",
      "from scipy import stats\n",
      "from scipy.spatial.distance import pdist, squareform\n",
      "import pandas\n",
      "import statsmodels.api as sm\n",
      "import operator\n",
      "import itertools\n",
      "import collections\n",
      "import functools\n",
      "from bisect import bisect_left, bisect_right\n",
      "import MySQLdb\n",
      "from datetime import datetime\n",
      "import pyfasta\n",
      "import pysam\n",
      "import pysamstats\n",
      "import petl\n",
      "import petl.interactive as etl\n",
      "etl.repr_index_header = True\n",
      "from petl.io import PopenSource, StringSource\n",
      "import petlx.all\n",
      "import vcf\n",
      "import vcfnp\n",
      "import sh\n",
      "import shelve\n",
      "\n",
      "sys.path.append('../../../src/python')\n",
      "import pyximport; pyximport.install()\n",
      "from optimisations import *\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Utility functions"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def log(*msg):\n",
      "    print >>sys.stderr, str(datetime.now()) + ' :: ' + ' '.join(map(str, msg))\n",
      "    sys.stderr.flush()\n",
      "    \n",
      "from IPython.display import display_html\n",
      "\n",
      "def pre(msg):\n",
      "    display_html('<pre style=\"line-height: 100%; font-size: .7em; display: inline-block; white-space: pre; background-color: #ff6; border: 0px solid #ddd; padding: 1px\">' + str(msg).replace('<', '&lt;') + '</pre>', raw=True)\n",
      "    \n",
      "def html(content):\n",
      "    display_html(content, raw=True)\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Constants"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "CROSSES = '3d7_hb3', 'hb3_dd2', '7g8_gb4'\n",
      "CHROMOSOMES = tuple('Pf3D7_%02d_v3' % n for n in range(1, 15))\n",
      "PARENT_RUNS = 'ERR019061', 'ERR019054', 'ERR012788', 'ERR012840', 'ERR027099', 'ERR027100'\n",
      "LABELS = {\n",
      "    '3d7_hb3': '3D7 x HB3', \n",
      "    'hb3_dd2': 'HB3 x Dd2', \n",
      "    '7g8_gb4': '7G8 x GB4',\n",
      "    'ERR019061': '3D7',\n",
      "    'ERR019054': 'HB3(1)',\n",
      "    'ERR012788': 'HB3(2)', \n",
      "    'ERR012840': 'Dd2',\n",
      "    'ERR027099': '7G8',\n",
      "    'ERR027100': 'GB4'\n",
      "}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "BASE_DIR = os.path.abspath('/data/plasmodium/pfalciparum/pf-crosses')\n",
      "DATA_DIR = os.path.join(BASE_DIR, 'data')\n",
      "RELEASE_DIR = os.path.join(DATA_DIR, 'release')\n",
      "PUBLIC_DIR = os.path.join(DATA_DIR, 'public/20141022')\n",
      "\n",
      "# public data\n",
      "GATK_CALLSET_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.gatk.final.npz')\n",
      "CORTEX_CALLSET_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.cortex.final.npz')\n",
      "COMBINED_VCF_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.combined.final.vcf.gz')\n",
      "COMBINED_CALLSET_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.combined.final.npz')\n",
      "CORTEX_RAW_VCF_FN_TEMPLATE = os.path.join(PUBLIC_DIR, '{cross}.cortex.raw.vcf.gz')\n",
      "BAM_FN_TEMPLATE = os.path.join(PUBLIC_DIR, 'bam/{run}.realigned.bam')\n",
      "COVERAGE_FN_TEMPLATE = os.path.join(PUBLIC_DIR, 'cnv/{run}.coverage_ext_binned.300.{chrom}.npy')\n",
      "CNV_FN_TEMPLATE = os.path.join(PUBLIC_DIR, 'cnv/{run}.cnv.300.{chrom}.npy')\n",
      "\n",
      "# reference data\n",
      "GENOME_DIR = os.path.join(DATA_DIR, 'genome/sanger/version3/September_2012')\n",
      "FASTA_FN = os.path.join(GENOME_DIR, 'Pf3D7_v3.fa')\n",
      "GFF_FN = os.path.join(GENOME_DIR, 'Pf3D7_v3.sorted.gff.gz')\n",
      "genome = pyfasta.Fasta(FASTA_FN)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dup_samples = {\n",
      "    '3d7_hb3': ['C01/PG0062-C/ERR019070',\n",
      "                'C02/PG0053-C/ERR019067',\n",
      "                'C02/PG0055-C/ERR019066',\n",
      "                'C02/PG0056-C/ERR019068'],\n",
      "    'hb3_dd2': [],\n",
      "    '7g8_gb4': ['AUD/PG0112-CW/ERR045639',\n",
      "                'JC9/PG0111-C/ERR029409',\n",
      "                'JE11/PG0100-CW/ERR045630',\n",
      "                'JF6/PG0079-CW/ERR045637',\n",
      "                'KB8/PG0104-CW/ERR045642',\n",
      "                'LA10/PG0086-CW/ERR045629',\n",
      "                'NIC/PG0095-C/ERR027107',\n",
      "                'QF5/PG0078-CW/ERR045638',\n",
      "                'XD8/PG0105-CW/ERR045628',\n",
      "                'XF12/PG0102-CW/ERR045635'],\n",
      "}\n",
      "\n",
      "\n",
      "excessive_recomb_samples = {\n",
      "    '3d7_hb3': [],\n",
      "    'hb3_dd2': ['SC01/PG0025-C/ERR019045'],\n",
      "    '7g8_gb4': ['D2/PG0094-CW/ERR045632'],\n",
      "}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Numpy utilities"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def interleave(a, b):\n",
      "    c = np.empty((a.size + b.size,), dtype=a.dtype)\n",
      "    c[0::2] = a\n",
      "    c[1::2] = b\n",
      "    return c\n",
      "\n",
      "\n",
      "def unpack_callset(callset):\n",
      "    V = callset['variants'].view(np.recarray)\n",
      "    C = callset['calldata'].view(np.recarray)\n",
      "    C2d = vcfnp.view2d(C).view(np.recarray)\n",
      "    return V, C, C2d\n",
      "\n",
      "\n",
      "def filter_variants(callset, query=None, verbose=True):\n",
      "    if query is None:\n",
      "        return callset\n",
      "    else:\n",
      "        V, C, _ = unpack_callset(callset)\n",
      "        if isinstance(query, basestring):\n",
      "            flt = numexpr.evaluate(query, local_dict=V)\n",
      "        elif callable(query):\n",
      "            flt = query(*unpack_callset(callset))\n",
      "        else:\n",
      "            flt = np.asarray(query)\n",
      "        indices = np.nonzero(flt)[0]\n",
      "        ret, exc, tot = indices.size, V.size-indices.size, V.size\n",
      "        if verbose:\n",
      "            log('filter variants: excluding %s (%.1f%%)' % (exc, exc*100./tot), 'retaining %s (%.1f%%)' % (ret, ret*100./tot), 'of', tot, 'variants')\n",
      "        V = np.take(V, indices, axis=0).view(np.recarray)\n",
      "        C = np.take(C, indices, axis=0).view(np.recarray)\n",
      "        return {'variants': V, 'calldata': C}\n",
      "\n",
      "\n",
      "def filter_calls(callset, query=None, verbose=True):\n",
      "    if query is None:\n",
      "        return callset\n",
      "    else:\n",
      "        V, C, _ = unpack_callset(callset)\n",
      "        C = C.copy()\n",
      "        C2d = vcfnp.view2d(C).view(np.recarray)\n",
      "\n",
      "        # filter calls\n",
      "        if isinstance(query, basestring):\n",
      "            flt = numexpr.evaluate(query, local_dict=C2d)\n",
      "        elif callable(query):\n",
      "            flt = query(*unpack_callset(callset))\n",
      "        else:\n",
      "            flt = np.asarray(query)\n",
      "        # only filter non-missing\n",
      "        excl = (C2d.genotype >= 0) & ~flt\n",
      "\n",
      "        tot, exc, ret = excl.size, np.count_nonzero(excl), np.count_nonzero(~excl)\n",
      "        if verbose:\n",
      "            log('filter calls: excluding %s (%.1f%%)' % (exc, exc*100./tot), 'retaining %s (%.1f%%)' % (ret, ret*100./tot), 'of', tot, 'calls')\n",
      "\n",
      "        C2d.genotype[excl] = -2\n",
      "        return {'variants': V, 'calldata': C}    \n",
      "    \n",
      "\n",
      "def filter_samples(callset, inclusions=None, exclusions=None, verbose=True):\n",
      "    if inclusions or exclusions:\n",
      "        V, C, _ = unpack_callset(callset)\n",
      "        samples = C.dtype.names\n",
      "        if inclusions:\n",
      "            exclusions = [s for s in samples if s not in inclusions]\n",
      "        elif exclusions:\n",
      "            inclusions = [s for s in samples if s not in exclusions]\n",
      "        if verbose:\n",
      "            log('filter samples: excluding', exclusions, 'including', inclusions)\n",
      "        C = C[inclusions].copy()\n",
      "        return {'variants': V, 'calldata': C}\n",
      "    else:\n",
      "        return callset\n",
      "    \n",
      "    \n",
      "def filter_callset(callset, \n",
      "                   variant_filter=None, \n",
      "                   call_filter=None, \n",
      "                   sample_inclusions=None, \n",
      "                   sample_exclusions=None,\n",
      "                   verbose=True):\n",
      "    \n",
      "    callset = filter_variants(callset, variant_filter, verbose=verbose)\n",
      "    \n",
      "    callset = filter_samples(callset, inclusions=sample_inclusions, exclusions=sample_exclusions, verbose=verbose)\n",
      "    \n",
      "    callset = filter_calls(callset, call_filter, verbose=verbose)\n",
      "    \n",
      "    return callset\n",
      "\n",
      "\n",
      "def load_callset(callset_fn, \n",
      "                 variant_filter=None, \n",
      "                 call_filter=None, \n",
      "                 sample_inclusions=None, \n",
      "                 sample_exclusions=None,\n",
      "                 verbose=True):\n",
      "    \n",
      "    if verbose:\n",
      "        log('loading', callset_fn)\n",
      "    callset = np.load(callset_fn)\n",
      "\n",
      "    callset = filter_callset(callset, \n",
      "                             variant_filter=variant_filter, \n",
      "                             call_filter=call_filter, \n",
      "                             sample_inclusions=sample_inclusions,\n",
      "                             sample_exclusions=sample_exclusions,\n",
      "                             verbose=verbose)\n",
      "    \n",
      "    return callset\n",
      "\n",
      "\n",
      "def filter_callsets(callsets,\n",
      "                    crosses=CROSSES,\n",
      "                    variant_filter=None, \n",
      "                    call_filter=None, \n",
      "                    sample_inclusions=dict(), \n",
      "                    sample_exclusions=dict(),\n",
      "                    verbose=True):\n",
      "    out = dict()\n",
      "    for cross in callsets.keys():\n",
      "        out[cross] = filter_callset(callsets[cross], \n",
      "                                    variant_filter=variant_filter,\n",
      "                                    call_filter=call_filter,\n",
      "                                    sample_inclusions=sample_inclusions.get(cross, None),\n",
      "                                    sample_exclusions=sample_exclusions.get(cross, None),\n",
      "                                    verbose=verbose)\n",
      "    return out\n",
      "    \n",
      "    \n",
      "def load_callsets(callset_fn_template,\n",
      "                  crosses=CROSSES,\n",
      "                  variant_filter=None, \n",
      "                  call_filter=None, \n",
      "                  sample_inclusions=dict(), \n",
      "                  sample_exclusions=dict(),\n",
      "                  verbose=True):\n",
      "    callsets = dict()\n",
      "    for cross in CROSSES:\n",
      "        callset_fn = callset_fn_template.format(cross=cross)\n",
      "        callsets[cross] = load_callset(callset_fn, \n",
      "                                       variant_filter=variant_filter,\n",
      "                                       call_filter=call_filter,\n",
      "                                       sample_inclusions=sample_inclusions.get(cross, None),\n",
      "                                       sample_exclusions=sample_exclusions.get(cross, None),\n",
      "                                       verbose=verbose)\n",
      "    return callsets\n",
      "    \n",
      "    \n",
      "def locate_variants(CHROM, POS, chrom, start=None, stop=None):\n",
      "    start_index, stop_index = bisect_left(CHROM, chrom), bisect_right(CHROM, chrom)\n",
      "    POS = POS[start_index:stop_index]\n",
      "    if stop is not None:\n",
      "        stop_index = start_index + bisect_right(POS, stop)\n",
      "    if start is not None:\n",
      "        start_index = start_index + bisect_left(POS, start)\n",
      "    return start_index, stop_index\n",
      "\n",
      "\n",
      "def min_haplen_calls(min_haplen):\n",
      "    def _f(v, c, c2d):\n",
      "        hl, hs, _ = haplotypes(v, c)\n",
      "        return hl >= min_haplen\n",
      "    return _f\n",
      "\n",
      "\n",
      "def combined_conf_calls(v, c, c2d):\n",
      "    return ((v.set[:,np.newaxis] == 'Cortex') & (c2d.GT_CONF >= 50)\n",
      "            | (v.set[:,np.newaxis] != 'Cortex') & (c2d.GQ >= 99))\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Tables"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def display_with_nrows(tbl, n=5, caption=''):\n",
      "    caption += ' (%s rows)' % tbl.nrows()\n",
      "    tbl.display(n, caption=caption)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "CACHE_DIR = os.path.join(PUBLIC_DIR, 'cache')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def etlcache(f):\n",
      "    fn = os.path.join(CACHE_DIR, f.__name__)\n",
      "    if not os.path.exists(fn):\n",
      "        etl.topickle(f(), fn)\n",
      "    return etl.frompickle(fn)\n",
      "    \n",
      "    \n",
      "def pdcache(f):\n",
      "    fn = os.path.join(CACHE_DIR, f.__name__)\n",
      "    if not os.path.exists(fn):\n",
      "        obj = f()\n",
      "        pandas.to_pickle(obj, fn)\n",
      "        return obj\n",
      "    else:\n",
      "        return pandas.read_pickle(fn)\n",
      "    \n",
      "    \n",
      "def nocache(f):\n",
      "    fn = os.path.join(CACHE_DIR, f.__name__)\n",
      "    if os.path.exists(fn):\n",
      "        os.remove(fn)\n",
      "    return f()\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "@etlcache\n",
      "def tbl_regions_1b():\n",
      "    return (etl\n",
      "        .fromtsv('../../../meta/regions-20130225.onebased.txt')\n",
      "        .pushheader(['chrom', 'start', 'stop', 'type'])\n",
      "        .convertnumbers()\n",
      "        .prefixheader('region_')\n",
      "        .addfield('region_size', lambda row: row.region_stop - row.region_start + 1)\n",
      "    )\n",
      "tbl_regions_1b"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<table class='petl'>\r\n",
        "<thead>\r\n",
        "<tr>\r\n",
        "<th>0|region_chrom</th>\r\n",
        "<th>1|region_start</th>\r\n",
        "<th>2|region_stop</th>\r\n",
        "<th>3|region_type</th>\r\n",
        "<th>4|region_size</th>\r\n",
        "</tr>\r\n",
        "</thead>\r\n",
        "<tbody>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td style='text-align: right'>1</td>\r\n",
        "<td style='text-align: right'>27336</td>\r\n",
        "<td>SubtelomericRepeat</td>\r\n",
        "<td style='text-align: right'>27336</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td style='text-align: right'>27337</td>\r\n",
        "<td style='text-align: right'>92900</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "<td style='text-align: right'>65564</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td style='text-align: right'>92901</td>\r\n",
        "<td style='text-align: right'>457931</td>\r\n",
        "<td>Core</td>\r\n",
        "<td style='text-align: right'>365031</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td style='text-align: right'>457932</td>\r\n",
        "<td style='text-align: right'>460311</td>\r\n",
        "<td>Centromere</td>\r\n",
        "<td style='text-align: right'>2380</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td style='text-align: right'>460312</td>\r\n",
        "<td style='text-align: right'>575900</td>\r\n",
        "<td>Core</td>\r\n",
        "<td style='text-align: right'>115589</td>\r\n",
        "</tr>\r\n",
        "</tbody>\r\n",
        "</table>\r\n",
        "<p><strong>...</strong></p>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "+------------------+------------------+-----------------+-----------------------------+-----------------+\n",
        "| '0|region_chrom' | '1|region_start' | '2|region_stop' | '3|region_type'             | '4|region_size' |\n",
        "+==================+==================+=================+=============================+=================+\n",
        "| 'Pf3D7_01_v3'    |                1 |           27336 | 'SubtelomericRepeat'        |           27336 |\n",
        "+------------------+------------------+-----------------+-----------------------------+-----------------+\n",
        "| 'Pf3D7_01_v3'    |            27337 |           92900 | 'SubtelomericHypervariable' |           65564 |\n",
        "+------------------+------------------+-----------------+-----------------------------+-----------------+\n",
        "| 'Pf3D7_01_v3'    |            92901 |          457931 | 'Core'                      |          365031 |\n",
        "+------------------+------------------+-----------------+-----------------------------+-----------------+\n",
        "| 'Pf3D7_01_v3'    |           457932 |          460311 | 'Centromere'                |            2380 |\n",
        "+------------------+------------------+-----------------+-----------------------------+-----------------+\n",
        "| 'Pf3D7_01_v3'    |           460312 |          575900 | 'Core'                      |          115589 |\n",
        "+------------------+------------------+-----------------+-----------------------------+-----------------+"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "@etlcache\n",
      "def tbl_features():\n",
      "    return (etl\n",
      "        .fromgff3(GFF_FN)\n",
      "        .unpackdict('attributes', ['ID', 'Parent', 'Name', 'previous_systematic_id'])\n",
      "        .rename({'seqid': 'chrom', 'end': 'stop', 'ID': 'id', 'Parent': 'parent_id', 'Name': 'name', 'previous_systematic_id': 'previous_id'})\n",
      "        .cutout('source', 'score', 'phase')\n",
      "        .addfield('length', lambda row: row.stop - row.start, index=4)\n",
      "        .intervalleftjoin(tbl_regions_1b, lfacet='chrom', rfacet='region_chrom', rstart='region_start', rstop='region_stop')\n",
      "        .cutout('region_chrom', 'region_start', 'region_stop')\n",
      "        .distinct('id')\n",
      "        .sort(key=('chrom', 'start', 'type'))\n",
      "        .prefixheader('feature_')\n",
      "    )\n",
      "tbl_features"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<table class='petl'>\r\n",
        "<thead>\r\n",
        "<tr>\r\n",
        "<th>0|feature_chrom</th>\r\n",
        "<th>1|feature_type</th>\r\n",
        "<th>2|feature_start</th>\r\n",
        "<th>3|feature_stop</th>\r\n",
        "<th>4|feature_length</th>\r\n",
        "<th>5|feature_strand</th>\r\n",
        "<th>6|feature_id</th>\r\n",
        "<th>7|feature_parent_id</th>\r\n",
        "<th>8|feature_name</th>\r\n",
        "<th>9|feature_previous_id</th>\r\n",
        "<th>10|feature_region_type</th>\r\n",
        "</tr>\r\n",
        "</thead>\r\n",
        "<tbody>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>repeat_region</td>\r\n",
        "<td style='text-align: right'>1</td>\r\n",
        "<td style='text-align: right'>360</td>\r\n",
        "<td style='text-align: right'>359</td>\r\n",
        "<td>+</td>\r\n",
        "<td>Pfalciparum_REP_20</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericRepeat</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>repeat_region</td>\r\n",
        "<td style='text-align: right'>361</td>\r\n",
        "<td style='text-align: right'>1418</td>\r\n",
        "<td style='text-align: right'>1057</td>\r\n",
        "<td>+</td>\r\n",
        "<td>Pfalciparum_REP_15</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericRepeat</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>repeat_region</td>\r\n",
        "<td style='text-align: right'>2160</td>\r\n",
        "<td style='text-align: right'>3858</td>\r\n",
        "<td style='text-align: right'>1698</td>\r\n",
        "<td>+</td>\r\n",
        "<td>Pfalciparum_REP_35</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericRepeat</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>repeat_region</td>\r\n",
        "<td style='text-align: right'>8856</td>\r\n",
        "<td style='text-align: right'>9021</td>\r\n",
        "<td style='text-align: right'>165</td>\r\n",
        "<td>+</td>\r\n",
        "<td>Pfalciparum_REP_5</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericRepeat</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>repeat_region</td>\r\n",
        "<td style='text-align: right'>9313</td>\r\n",
        "<td style='text-align: right'>9529</td>\r\n",
        "<td style='text-align: right'>216</td>\r\n",
        "<td>+</td>\r\n",
        "<td>Pfalciparum_REP_25</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericRepeat</td>\r\n",
        "</tr>\r\n",
        "</tbody>\r\n",
        "</table>\r\n",
        "<p><strong>...</strong></p>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 9,
       "text": [
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+----------------------+-----------------------+------------------+-------------------------+--------------------------+\n",
        "| '0|feature_chrom' | '1|feature_type' | '2|feature_start' | '3|feature_stop' | '4|feature_length' | '5|feature_strand' | '6|feature_id'       | '7|feature_parent_id' | '8|feature_name' | '9|feature_previous_id' | '10|feature_region_type' |\n",
        "+===================+==================+===================+==================+====================+====================+======================+=======================+==================+=========================+==========================+\n",
        "| 'Pf3D7_01_v3'     | 'repeat_region'  |                 1 |              360 |                359 | '+'                | 'Pfalciparum_REP_20' | None                  | None             | None                    | 'SubtelomericRepeat'     |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+----------------------+-----------------------+------------------+-------------------------+--------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'repeat_region'  |               361 |             1418 |               1057 | '+'                | 'Pfalciparum_REP_15' | None                  | None             | None                    | 'SubtelomericRepeat'     |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+----------------------+-----------------------+------------------+-------------------------+--------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'repeat_region'  |              2160 |             3858 |               1698 | '+'                | 'Pfalciparum_REP_35' | None                  | None             | None                    | 'SubtelomericRepeat'     |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+----------------------+-----------------------+------------------+-------------------------+--------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'repeat_region'  |              8856 |             9021 |                165 | '+'                | 'Pfalciparum_REP_5'  | None                  | None             | None                    | 'SubtelomericRepeat'     |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+----------------------+-----------------------+------------------+-------------------------+--------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'repeat_region'  |              9313 |             9529 |                216 | '+'                | 'Pfalciparum_REP_25' | None                  | None             | None                    | 'SubtelomericRepeat'     |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+----------------------+-----------------------+------------------+-------------------------+--------------------------+"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "lkp_feature = tbl_features.recordlookupone('feature_id')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "@etlcache\n",
      "def tbl_genes():\n",
      "    return (\n",
      "        tbl_features\n",
      "        .selectin('feature_type', ['gene', 'pseudogene'])\n",
      "        .cutout('feature_parent_id')\n",
      "    )\n",
      "tbl_genes"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<table class='petl'>\r\n",
        "<thead>\r\n",
        "<tr>\r\n",
        "<th>0|feature_chrom</th>\r\n",
        "<th>1|feature_type</th>\r\n",
        "<th>2|feature_start</th>\r\n",
        "<th>3|feature_stop</th>\r\n",
        "<th>4|feature_length</th>\r\n",
        "<th>5|feature_strand</th>\r\n",
        "<th>6|feature_id</th>\r\n",
        "<th>7|feature_name</th>\r\n",
        "<th>8|feature_previous_id</th>\r\n",
        "<th>9|feature_region_type</th>\r\n",
        "</tr>\r\n",
        "</thead>\r\n",
        "<tbody>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>gene</td>\r\n",
        "<td style='text-align: right'>29510</td>\r\n",
        "<td style='text-align: right'>37126</td>\r\n",
        "<td style='text-align: right'>7616</td>\r\n",
        "<td>+</td>\r\n",
        "<td>PF3D7_0100100</td>\r\n",
        "<td>VAR</td>\r\n",
        "<td>PFA0005w</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>gene</td>\r\n",
        "<td style='text-align: right'>38982</td>\r\n",
        "<td style='text-align: right'>40207</td>\r\n",
        "<td style='text-align: right'>1225</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100200</td>\r\n",
        "<td>RIF</td>\r\n",
        "<td>PFA0010c</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>gene</td>\r\n",
        "<td style='text-align: right'>42367</td>\r\n",
        "<td style='text-align: right'>46507</td>\r\n",
        "<td style='text-align: right'>4140</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100300</td>\r\n",
        "<td>None</td>\r\n",
        "<td>PFA0015c</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>gene</td>\r\n",
        "<td style='text-align: right'>50363</td>\r\n",
        "<td style='text-align: right'>51636</td>\r\n",
        "<td style='text-align: right'>1273</td>\r\n",
        "<td>+</td>\r\n",
        "<td>PF3D7_0100400</td>\r\n",
        "<td>RIF</td>\r\n",
        "<td>PFA0020w</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>pseudogene</td>\r\n",
        "<td style='text-align: right'>53169</td>\r\n",
        "<td style='text-align: right'>53280</td>\r\n",
        "<td style='text-align: right'>111</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100500</td>\r\n",
        "<td>var pseudogene, exon 1</td>\r\n",
        "<td>PFA0025c</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "</tbody>\r\n",
        "</table>\r\n",
        "<p><strong>...</strong></p>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 11,
       "text": [
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-----------------+--------------------------+-------------------------+-----------------------------+\n",
        "| '0|feature_chrom' | '1|feature_type' | '2|feature_start' | '3|feature_stop' | '4|feature_length' | '5|feature_strand' | '6|feature_id'  | '7|feature_name'         | '8|feature_previous_id' | '9|feature_region_type'     |\n",
        "+===================+==================+===================+==================+====================+====================+=================+==========================+=========================+=============================+\n",
        "| 'Pf3D7_01_v3'     | 'gene'           |             29510 |            37126 |               7616 | '+'                | 'PF3D7_0100100' | 'VAR'                    | 'PFA0005w'              | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-----------------+--------------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'gene'           |             38982 |            40207 |               1225 | '-'                | 'PF3D7_0100200' | 'RIF'                    | 'PFA0010c'              | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-----------------+--------------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'gene'           |             42367 |            46507 |               4140 | '-'                | 'PF3D7_0100300' | None                     | 'PFA0015c'              | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-----------------+--------------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'gene'           |             50363 |            51636 |               1273 | '+'                | 'PF3D7_0100400' | 'RIF'                    | 'PFA0020w'              | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-----------------+--------------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'pseudogene'     |             53169 |            53280 |                111 | '-'                | 'PF3D7_0100500' | 'var pseudogene, exon 1' | 'PFA0025c'              | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-----------------+--------------------------+-------------------------+-----------------------------+"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "@etlcache\n",
      "def tbl_intergenic():\n",
      "    return (\n",
      "        tbl_genes\n",
      "        .addfieldusingcontext('next', lambda prv, cur, nxt: (nxt.feature_start, nxt.feature_id, nxt.feature_strand) if nxt and cur.feature_chrom == nxt.feature_chrom else None)\n",
      "        .notnone('next')\n",
      "        .unpack('next', ['next_start', 'next_id', 'next_strand'])\n",
      "        .convert('feature_id', lambda v, row: '%s_%s' % (row.feature_id, row.next_id), pass_row=True)\n",
      "        .cutout('feature_start', 'feature_length', 'feature_region_type', 'next_id')\n",
      "        .rename({'feature_stop': 'feature_start', 'next_start': 'feature_stop'})\n",
      "        .movefield('feature_stop', 3)\n",
      "        .addfield('feature_length', lambda row: row.feature_stop - row.feature_start, index=4)\n",
      "        .convert('feature_type', lambda v, row: 'intergenic %s/%s' % (row.feature_strand, row.next_strand), pass_row=True)\n",
      "        # reclassify based on expected number of promoters\n",
      "        .convert('feature_type', {'intergenic +/-': 'intergenic_0', 'intergenic -/+': 'intergenic_2', 'intergenic +/+': 'intergenic_1', 'intergenic -/-': 'intergenic_1'})\n",
      "        .cutout('feature_strand', 'next_strand', 'feature_name', 'feature_previous_id')\n",
      "        .intervalleftjoin(tbl_regions_1b, \n",
      "                          lfacet='feature_chrom', lstart='feature_start', lstop='feature_stop',\n",
      "                          rfacet='region_chrom', rstart='region_start', rstop='region_stop')\n",
      "        .rename('region_type', 'feature_region_type')\n",
      "        .cutout('region_chrom', 'region_start', 'region_stop')\n",
      "        .gt('feature_length', 0)\n",
      "    )\n",
      "\n",
      "tbl_intergenic"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<table class='petl'>\r\n",
        "<thead>\r\n",
        "<tr>\r\n",
        "<th>0|feature_chrom</th>\r\n",
        "<th>1|feature_type</th>\r\n",
        "<th>2|feature_start</th>\r\n",
        "<th>3|feature_stop</th>\r\n",
        "<th>4|feature_length</th>\r\n",
        "<th>5|feature_id</th>\r\n",
        "<th>6|feature_region_type</th>\r\n",
        "</tr>\r\n",
        "</thead>\r\n",
        "<tbody>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intergenic_0</td>\r\n",
        "<td style='text-align: right'>37126</td>\r\n",
        "<td style='text-align: right'>38982</td>\r\n",
        "<td style='text-align: right'>1856</td>\r\n",
        "<td>PF3D7_0100100_PF3D7_0100200</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intergenic_1</td>\r\n",
        "<td style='text-align: right'>40207</td>\r\n",
        "<td style='text-align: right'>42367</td>\r\n",
        "<td style='text-align: right'>2160</td>\r\n",
        "<td>PF3D7_0100200_PF3D7_0100300</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intergenic_2</td>\r\n",
        "<td style='text-align: right'>46507</td>\r\n",
        "<td style='text-align: right'>50363</td>\r\n",
        "<td style='text-align: right'>3856</td>\r\n",
        "<td>PF3D7_0100300_PF3D7_0100400</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intergenic_0</td>\r\n",
        "<td style='text-align: right'>51636</td>\r\n",
        "<td style='text-align: right'>53169</td>\r\n",
        "<td style='text-align: right'>1533</td>\r\n",
        "<td>PF3D7_0100400_PF3D7_0100500</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intergenic_1</td>\r\n",
        "<td style='text-align: right'>53280</td>\r\n",
        "<td style='text-align: right'>53778</td>\r\n",
        "<td style='text-align: right'>498</td>\r\n",
        "<td>PF3D7_0100500_PF3D7_0100600</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "</tbody>\r\n",
        "</table>\r\n",
        "<p><strong>...</strong></p>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "+-------------------+------------------+-------------------+------------------+--------------------+-------------------------------+-----------------------------+\n",
        "| '0|feature_chrom' | '1|feature_type' | '2|feature_start' | '3|feature_stop' | '4|feature_length' | '5|feature_id'                | '6|feature_region_type'     |\n",
        "+===================+==================+===================+==================+====================+===============================+=============================+\n",
        "| 'Pf3D7_01_v3'     | 'intergenic_0'   |             37126 |            38982 |               1856 | 'PF3D7_0100100_PF3D7_0100200' | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+-------------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'intergenic_1'   |             40207 |            42367 |               2160 | 'PF3D7_0100200_PF3D7_0100300' | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+-------------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'intergenic_2'   |             46507 |            50363 |               3856 | 'PF3D7_0100300_PF3D7_0100400' | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+-------------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'intergenic_0'   |             51636 |            53169 |               1533 | 'PF3D7_0100400_PF3D7_0100500' | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+-------------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'intergenic_1'   |             53280 |            53778 |                498 | 'PF3D7_0100500_PF3D7_0100600' | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+-------------------------------+-----------------------------+"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "@etlcache\n",
      "def tbl_exons():\n",
      "    return (\n",
      "        tbl_features\n",
      "        .eq('feature_type', 'CDS')\n",
      "        .gt('feature_length', 0)\n",
      "    )\n",
      "\n",
      "tbl_exons"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<table class='petl'>\r\n",
        "<thead>\r\n",
        "<tr>\r\n",
        "<th>0|feature_chrom</th>\r\n",
        "<th>1|feature_type</th>\r\n",
        "<th>2|feature_start</th>\r\n",
        "<th>3|feature_stop</th>\r\n",
        "<th>4|feature_length</th>\r\n",
        "<th>5|feature_strand</th>\r\n",
        "<th>6|feature_id</th>\r\n",
        "<th>7|feature_parent_id</th>\r\n",
        "<th>8|feature_name</th>\r\n",
        "<th>9|feature_previous_id</th>\r\n",
        "<th>10|feature_region_type</th>\r\n",
        "</tr>\r\n",
        "</thead>\r\n",
        "<tbody>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>CDS</td>\r\n",
        "<td style='text-align: right'>29510</td>\r\n",
        "<td style='text-align: right'>34762</td>\r\n",
        "<td style='text-align: right'>5252</td>\r\n",
        "<td>+</td>\r\n",
        "<td>PF3D7_0100100.1:exon:1</td>\r\n",
        "<td>PF3D7_0100100.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>CDS</td>\r\n",
        "<td style='text-align: right'>35888</td>\r\n",
        "<td style='text-align: right'>37126</td>\r\n",
        "<td style='text-align: right'>1238</td>\r\n",
        "<td>+</td>\r\n",
        "<td>PF3D7_0100100.1:exon:2</td>\r\n",
        "<td>PF3D7_0100100.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>CDS</td>\r\n",
        "<td style='text-align: right'>38982</td>\r\n",
        "<td style='text-align: right'>39923</td>\r\n",
        "<td style='text-align: right'>941</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100200.1:exon:1</td>\r\n",
        "<td>PF3D7_0100200.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>CDS</td>\r\n",
        "<td style='text-align: right'>40154</td>\r\n",
        "<td style='text-align: right'>40207</td>\r\n",
        "<td style='text-align: right'>53</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100200.1:exon:2</td>\r\n",
        "<td>PF3D7_0100200.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>CDS</td>\r\n",
        "<td style='text-align: right'>42367</td>\r\n",
        "<td style='text-align: right'>43617</td>\r\n",
        "<td style='text-align: right'>1250</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100300.1:exon:1</td>\r\n",
        "<td>PF3D7_0100300.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "</tbody>\r\n",
        "</table>\r\n",
        "<p><strong>...</strong></p>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 13,
       "text": [
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+--------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| '0|feature_chrom' | '1|feature_type' | '2|feature_start' | '3|feature_stop' | '4|feature_length' | '5|feature_strand' | '6|feature_id'           | '7|feature_parent_id' | '8|feature_name' | '9|feature_previous_id' | '10|feature_region_type'    |\n",
        "+===================+==================+===================+==================+====================+====================+==========================+=======================+==================+=========================+=============================+\n",
        "| 'Pf3D7_01_v3'     | 'CDS'            |             29510 |            34762 |               5252 | '+'                | 'PF3D7_0100100.1:exon:1' | 'PF3D7_0100100.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+--------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'CDS'            |             35888 |            37126 |               1238 | '+'                | 'PF3D7_0100100.1:exon:2' | 'PF3D7_0100100.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+--------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'CDS'            |             38982 |            39923 |                941 | '-'                | 'PF3D7_0100200.1:exon:1' | 'PF3D7_0100200.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+--------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'CDS'            |             40154 |            40207 |                 53 | '-'                | 'PF3D7_0100200.1:exon:2' | 'PF3D7_0100200.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+--------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'CDS'            |             42367 |            43617 |               1250 | '-'                | 'PF3D7_0100300.1:exon:1' | 'PF3D7_0100300.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+--------------------------+-----------------------+------------------+-------------------------+-----------------------------+"
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "@etlcache\n",
      "def tbl_introns():\n",
      "    return (\n",
      "        tbl_exons\n",
      "        .addfieldusingcontext('next', \n",
      "                              lambda prv, cur, nxt: (nxt.feature_start, nxt.feature_id) if nxt and cur.feature_parent_id == nxt.feature_parent_id else None)\n",
      "        .notnone('next')\n",
      "        .unpack('next', ['next_start', 'next_id'])\n",
      "        .convert('feature_id', lambda v, row: '%s_%s' % (row.feature_id, row.next_id), pass_row=True)\n",
      "        .cutout('feature_start', 'feature_length', 'feature_region_type', 'next_id')\n",
      "        .rename({'feature_stop': 'feature_start', 'next_start': 'feature_stop'})\n",
      "        .movefield('feature_stop', 3)\n",
      "        .addfield('feature_length', lambda row: row.feature_stop - row.feature_start, index=4)\n",
      "        .convert('feature_type', lambda v: 'intron')\n",
      "        .intervalleftjoin(tbl_regions_1b, \n",
      "                          lfacet='feature_chrom', lstart='feature_start', lstop='feature_stop',\n",
      "                          rfacet='region_chrom', rstart='region_start', rstop='region_stop')\n",
      "        .rename('region_type', 'feature_region_type')\n",
      "        .cutout('region_chrom', 'region_start', 'region_stop')\n",
      "    )\n",
      "\n",
      "tbl_introns"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<table class='petl'>\r\n",
        "<thead>\r\n",
        "<tr>\r\n",
        "<th>0|feature_chrom</th>\r\n",
        "<th>1|feature_type</th>\r\n",
        "<th>2|feature_start</th>\r\n",
        "<th>3|feature_stop</th>\r\n",
        "<th>4|feature_length</th>\r\n",
        "<th>5|feature_strand</th>\r\n",
        "<th>6|feature_id</th>\r\n",
        "<th>7|feature_parent_id</th>\r\n",
        "<th>8|feature_name</th>\r\n",
        "<th>9|feature_previous_id</th>\r\n",
        "<th>10|feature_region_type</th>\r\n",
        "</tr>\r\n",
        "</thead>\r\n",
        "<tbody>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intron</td>\r\n",
        "<td style='text-align: right'>34762</td>\r\n",
        "<td style='text-align: right'>35888</td>\r\n",
        "<td style='text-align: right'>1126</td>\r\n",
        "<td>+</td>\r\n",
        "<td>PF3D7_0100100.1:exon:1_PF3D7_0100100.1:exon:2</td>\r\n",
        "<td>PF3D7_0100100.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intron</td>\r\n",
        "<td style='text-align: right'>39923</td>\r\n",
        "<td style='text-align: right'>40154</td>\r\n",
        "<td style='text-align: right'>231</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100200.1:exon:1_PF3D7_0100200.1:exon:2</td>\r\n",
        "<td>PF3D7_0100200.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intron</td>\r\n",
        "<td style='text-align: right'>43617</td>\r\n",
        "<td style='text-align: right'>43775</td>\r\n",
        "<td style='text-align: right'>158</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100300.1:exon:1_PF3D7_0100300.1:exon:2</td>\r\n",
        "<td>PF3D7_0100300.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intron</td>\r\n",
        "<td style='text-align: right'>50416</td>\r\n",
        "<td style='text-align: right'>50572</td>\r\n",
        "<td style='text-align: right'>156</td>\r\n",
        "<td>+</td>\r\n",
        "<td>PF3D7_0100400.1:exon:1_PF3D7_0100400.1:exon:2</td>\r\n",
        "<td>PF3D7_0100400.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "<tr>\r\n",
        "<td>Pf3D7_01_v3</td>\r\n",
        "<td>intron</td>\r\n",
        "<td style='text-align: right'>54788</td>\r\n",
        "<td style='text-align: right'>54938</td>\r\n",
        "<td style='text-align: right'>150</td>\r\n",
        "<td>-</td>\r\n",
        "<td>PF3D7_0100600.1:exon:1_PF3D7_0100600.1:exon:2</td>\r\n",
        "<td>PF3D7_0100600.1</td>\r\n",
        "<td>None</td>\r\n",
        "<td>None</td>\r\n",
        "<td>SubtelomericHypervariable</td>\r\n",
        "</tr>\r\n",
        "</tbody>\r\n",
        "</table>\r\n",
        "<p><strong>...</strong></p>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 14,
       "text": [
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-------------------------------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| '0|feature_chrom' | '1|feature_type' | '2|feature_start' | '3|feature_stop' | '4|feature_length' | '5|feature_strand' | '6|feature_id'                                  | '7|feature_parent_id' | '8|feature_name' | '9|feature_previous_id' | '10|feature_region_type'    |\n",
        "+===================+==================+===================+==================+====================+====================+=================================================+=======================+==================+=========================+=============================+\n",
        "| 'Pf3D7_01_v3'     | 'intron'         |             34762 |            35888 |               1126 | '+'                | 'PF3D7_0100100.1:exon:1_PF3D7_0100100.1:exon:2' | 'PF3D7_0100100.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-------------------------------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'intron'         |             39923 |            40154 |                231 | '-'                | 'PF3D7_0100200.1:exon:1_PF3D7_0100200.1:exon:2' | 'PF3D7_0100200.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-------------------------------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'intron'         |             43617 |            43775 |                158 | '-'                | 'PF3D7_0100300.1:exon:1_PF3D7_0100300.1:exon:2' | 'PF3D7_0100300.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-------------------------------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'intron'         |             50416 |            50572 |                156 | '+'                | 'PF3D7_0100400.1:exon:1_PF3D7_0100400.1:exon:2' | 'PF3D7_0100400.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-------------------------------------------------+-----------------------+------------------+-------------------------+-----------------------------+\n",
        "| 'Pf3D7_01_v3'     | 'intron'         |             54788 |            54938 |                150 | '-'                | 'PF3D7_0100600.1:exon:1_PF3D7_0100600.1:exon:2' | 'PF3D7_0100600.1'     | None             | None                    | 'SubtelomericHypervariable' |\n",
        "+-------------------+------------------+-------------------+------------------+--------------------+--------------------+-------------------------------------------------+-----------------------+------------------+-------------------------+-----------------------------+"
       ]
      }
     ],
     "prompt_number": 14
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Plotting"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_features(chrom, start=None, stop=None):\n",
      "    if None not in {start, stop}:\n",
      "        region = '%s:%s-%s' % (chrom, start, stop)\n",
      "    else:\n",
      "        region = chrom\n",
      "    tbl_features = (etl\n",
      "        .fromgff3(GFF_FN, region)\n",
      "        .cutout('source', 'score', 'phase')\n",
      "        .unpackdict('attributes', ['ID', 'Name', 'previous_systematic_id'], includeoriginal=True)\n",
      "    )\n",
      "    tbl_features = tbl_features.setheader(['feature_' + str(f).lower() for f in tbl_features.header()])\n",
      "    return tbl_features\n",
      "\n",
      "\n",
      "prog_hv = re.compile('var|rif|stev', flags=re.I) \n",
      "\n",
      "\n",
      "def plot_genes(ax, chrom, start=0, stop=None, label_named=False, exclude_labels=[], width=.3, label_ytextpad=.1,\n",
      "               normal_color='k', hypervariable_color='r', cds_color='gray', divider_linewidth=1,\n",
      "               annotate_genes=[], annotate_xtext=0, annotate_ytext_fwd=6, annotate_ytext_rev=8, \n",
      "               annotate_arrowprops=None, annotate_bbox=None, \n",
      "               show_cds=False, gene_labels=dict()):\n",
      "    if stop is None:\n",
      "        stop = len(genome[chrom])\n",
      "    tbl_local_features = get_features(chrom, start, stop)\n",
      "    tbl_local_genes = tbl_local_features.selectin('feature_type', {'gene', 'pseudogene'})\n",
      "    tbl_local_cds = tbl_local_features.eq('feature_type', 'CDS')\n",
      "    \n",
      "    # divider\n",
      "    ax.plot([start, stop], [.5, .5], 'k-', linewidth=divider_linewidth)\n",
      "    ax.set_ylim(0, 1)\n",
      "    \n",
      "    # positive strand\n",
      "    tbl_fwd_genes = tbl_local_genes.eq('feature_strand', '+')\n",
      "    xranges = [(r.feature_start, r.feature_end-r.feature_start) for r in tbl_fwd_genes.records()]\n",
      "    fc = [hypervariable_color if r.feature_name is not None and prog_hv.search(r.feature_name) is not None else normal_color \n",
      "          for r in tbl_fwd_genes.records()]\n",
      "    ax.broken_barh(xranges, (.5, width), color=fc)\n",
      "    if show_cds:\n",
      "        tbl_fwd_cds = tbl_local_cds.eq('feature_strand', '+')\n",
      "        xranges = [(r.feature_start, r.feature_end-r.feature_start) for r in tbl_fwd_cds.records()]\n",
      "        ax.broken_barh(xranges, (.5, width), color=cds_color)\n",
      "        \n",
      "    \n",
      "    # negative strand\n",
      "    tbl_rev_genes = tbl_local_genes.eq('feature_strand', '-')\n",
      "    xranges = [(r.feature_start, r.feature_end-r.feature_start) for r in tbl_rev_genes.records()]\n",
      "    fc = [hypervariable_color if r.feature_name is not None and prog_hv.search(r.feature_name) is not None else normal_color\n",
      "          for r in tbl_rev_genes.records()]\n",
      "    ax.broken_barh(xranges, (.5-width, width), color=fc)\n",
      "    if show_cds:\n",
      "        tbl_rev_cds = tbl_local_cds.eq('feature_strand', '-')\n",
      "        xranges = [(r.feature_start, r.feature_end-r.feature_start) for r in tbl_rev_cds.records()]\n",
      "        ax.broken_barh(xranges, (.5-width, width), color=cds_color)\n",
      "    \n",
      "    for s in 'top', 'left', 'right':\n",
      "        ax.spines[s].set_visible(False)\n",
      "    ax.set_xlim(start, stop)\n",
      "    ax.set_yticks([])\n",
      "    ax.xaxis.tick_bottom()\n",
      "\n",
      "    # label named genes\n",
      "    if label_named:\n",
      "        for g in tbl_local_genes.records():\n",
      "            if g.feature_name is not None and g.feature_name not in exclude_labels:\n",
      "                x = g.feature_start\n",
      "                y = .5 + width + label_ytextpad if g.feature_strand == '+' else .5 - width - label_ytextpad\n",
      "                va = 'bottom' if g.feature_strand == '+' else 'top'\n",
      "                l = g.feature_name\n",
      "                ax.text(x, y, l, rotation=0, ha='left', va=va, transform=ax.transData)\n",
      "                \n",
      "    # annotate specific genes\n",
      "    if annotate_arrowprops is None:\n",
      "        annotate_arrowprops = dict(facecolor='k', arrowstyle='-', shrinkA=0, shrinkB=0)\n",
      "    if annotate_bbox is None:\n",
      "        annotate_bbox = dict(facecolor='w', alpha=.9)\n",
      "    for id in annotate_genes:\n",
      "        g = lkp_feature[id]\n",
      "        gene_id = g['feature_id']\n",
      "        l = gene_labels[gene_id] if gene_id in gene_labels else g['feature_name'] if g['feature_name'] is not None else gene_id\n",
      "        if g['feature_strand'] == '+':\n",
      "            ax.annotate(l, xy=(g['feature_start'], .5), ha='left', xytext=(annotate_xtext, annotate_ytext_fwd), \n",
      "                        xycoords='data', textcoords='offset points', \n",
      "                        arrowprops=annotate_arrowprops,\n",
      "                        bbox=annotate_bbox)\n",
      "        elif g['feature_strand'] == '-':\n",
      "            ax.annotate(l, xy=(g['feature_start'], .5), ha='left', xytext=(annotate_xtext, annotate_ytext_rev), \n",
      "                        xycoords='data', textcoords='offset points', \n",
      "                        arrowprops=annotate_arrowprops,\n",
      "                        bbox=annotate_bbox)\n",
      "    \n",
      "    \n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 're' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-2-56aeec77aa91>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     14\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 15\u001b[1;33m \u001b[0mprog_hv\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mre\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcompile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'var|rif|stev'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mre\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mI\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     16\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     17\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;31mNameError\u001b[0m: name 're' is not defined"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Recombination"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def marker_span(callset, chrom):\n",
      "    V, _, _ = unpack_callset(callset)\n",
      "    start_index, stop_index = locate_variants(V.CHROM, V.POS, chrom)\n",
      "    POS = V.POS[start_index:stop_index]\n",
      "    x = POS[-1] - POS[0]\n",
      "    return x\n",
      "\n",
      "\n",
      "def marker_offset(callset, chrom):\n",
      "    V, _, _ = unpack_callset(callset)\n",
      "    start_index, stop_index = locate_variants(V.CHROM, V.POS, chrom)\n",
      "    POS = V.POS[start_index:stop_index]\n",
      "    return POS[0]\n",
      "\n",
      "\n",
      "def locate_flanking_markers(CHROM, POS, chrom, pos):\n",
      "    chrom_start_index, chrom_stop_index = bisect_left(CHROM, chrom), bisect_right(CHROM, chrom)\n",
      "    POS = POS[chrom_start_index:chrom_stop_index]\n",
      "    \n",
      "    start_index = bisect_left(POS, pos) - 1\n",
      "    stop_index = bisect_right(POS, pos)\n",
      "    \n",
      "    if start_index < 0:\n",
      "        start_index = None\n",
      "    else:\n",
      "        start_index += chrom_start_index\n",
      "        \n",
      "    if stop_index >= len(POS):\n",
      "        stop_index = None\n",
      "    else:\n",
      "        stop_index += chrom_start_index\n",
      "    \n",
      "    return start_index, stop_index\n",
      "    \n",
      "\n",
      "def simulate_random_uniform_crossovers(n_events, callset, chromosomes=CHROMOSOMES):\n",
      "    \n",
      "    # table of results\n",
      "    tbl = [['chrom', 'co_pos', 'co_pos_mid', 'co_pos_min', 'co_pos_max', 'co_pos_range']]\n",
      "    \n",
      "    # utility variables\n",
      "    V = callset['variants']\n",
      "    marker_spans = [marker_span(callset, chrom) for chrom in chromosomes]\n",
      "    marker_offsets = [marker_offset(callset, chrom) for chrom in chromosomes]\n",
      "    cumulative_marker_span = [np.sum(marker_spans[:i+1]) for i in range(len(marker_spans))]\n",
      "    total_marker_span = np.sum(marker_spans)\n",
      "    \n",
      "    # generate random integers over the total marker span\n",
      "    X = np.random.randint(low=0, high=total_marker_span, size=n_events)\n",
      "    \n",
      "    # map each event onto a chromosome and position\n",
      "    for x in X:\n",
      "\n",
      "        # which chromosome?\n",
      "        chrom_index = bisect_left(cumulative_marker_span, x)\n",
      "        chrom = chromosomes[chrom_index]\n",
      "\n",
      "        # which position?\n",
      "        if chrom_index > 0:\n",
      "            pos_relative = x - cumulative_marker_span[chrom_index-1]\n",
      "        else:\n",
      "            pos_relative = x\n",
      "        pos = pos_relative + marker_offsets[chrom_index]\n",
      "        \n",
      "        # find flanking markers\n",
      "        lidx, ridx = locate_flanking_markers(V.CHROM, V.POS, chrom, pos)\n",
      "\n",
      "        pos_min = V.POS[lidx]\n",
      "        pos_max = V.POS[ridx]\n",
      "        pos_mid = (pos_min + pos_max)/2\n",
      "        pos_range = pos_max - pos_min\n",
      "        \n",
      "        row = [chrom, pos, pos_mid, pos_min, pos_max, pos_range]\n",
      "        tbl += [row]\n",
      "        \n",
      "    return etl.wrap(tbl)\n",
      "\n",
      "\n",
      "def simulate_random_uniform_non_crossovers(n_events, callset, phi, chromosomes=CHROMOSOMES):\n",
      "    \n",
      "    # table of results\n",
      "    tbl = [['chrom', \n",
      "            'tract_start', \n",
      "            'tract_stop', \n",
      "            'tract_length',\n",
      "            'tract_start_min', \n",
      "            'tract_start_mid', \n",
      "            'tract_start_max', \n",
      "            'tract_stop_min', \n",
      "            'tract_stop_mid', \n",
      "            'tract_stop_max', \n",
      "            'tract_length_min', \n",
      "            'tract_length_mid', \n",
      "            'tract_length_max', \n",
      "            'tract_support']]\n",
      "    \n",
      "    # utility variables\n",
      "    V = callset['variants']\n",
      "    marker_spans = [marker_span(callset, chrom) for chrom in chromosomes]\n",
      "    marker_offsets = [marker_offset(callset, chrom) for chrom in chromosomes]\n",
      "    cumulative_marker_span = [np.sum(marker_spans[:i+1]) for i in range(len(marker_spans))]\n",
      "    total_marker_span = np.sum(marker_spans)\n",
      "    \n",
      "    # generate random integers over the total marker span\n",
      "    X = np.random.randint(low=0, high=total_marker_span, size=n_events)\n",
      "\n",
      "    # generate random geometric tract lengths\n",
      "    lengths = stats.geom.rvs(1-phi, size=n_events).astype('int32')\n",
      "    \n",
      "    # map each event onto a chromosome and position\n",
      "    for x, length in zip(X, lengths):\n",
      "\n",
      "        try:\n",
      "            # which chromosome?\n",
      "            chrom_index = bisect_left(cumulative_marker_span, x)\n",
      "            chrom = chromosomes[chrom_index]\n",
      "\n",
      "            # which start position?\n",
      "            if chrom_index > 0:\n",
      "                pos_relative = x - cumulative_marker_span[chrom_index-1]\n",
      "            else:\n",
      "                pos_relative = x\n",
      "            start = pos_relative + marker_offsets[chrom_index]\n",
      "            stop = start + length\n",
      "\n",
      "            # find flanking markers\n",
      "            start_lidx, start_ridx = locate_flanking_markers(V.CHROM, V.POS, chrom, start)\n",
      "            stop_lidx, stop_ridx = locate_flanking_markers(V.CHROM, V.POS, chrom, stop)\n",
      "            \n",
      "            if None in {start_lidx, start_ridx, stop_lidx, stop_ridx}:\n",
      "                \n",
      "                # falls off edge of marker span\n",
      "                support = start_min = start_max = start_mid = stop_min = stop_max = stop_mid = length_min = length_max = length_mid = None\n",
      "\n",
      "            else:\n",
      "\n",
      "                support = stop_ridx - start_ridx\n",
      "\n",
      "                # do we even observe the event?\n",
      "                if support == 0:\n",
      "                    \n",
      "                    start_min = start_max = start_mid = stop_min = stop_max = stop_mid = length_min = length_max = length_mid = None\n",
      "\n",
      "                else:\n",
      "\n",
      "                    start_min = V.POS[start_lidx]\n",
      "                    start_max = V.POS[start_ridx]\n",
      "                    start_mid = (start_min + start_max)/2\n",
      "                    stop_min = V.POS[stop_lidx]\n",
      "                    stop_max = V.POS[stop_ridx]\n",
      "                    stop_mid = (stop_min + stop_max)/2\n",
      "                    length_min = stop_min - start_max\n",
      "                    length_max = stop_max - start_min\n",
      "                    length_mid = (length_min + length_max)/2\n",
      "\n",
      "            row = [chrom, \n",
      "                   start, \n",
      "                   stop, \n",
      "                   length,\n",
      "                   start_min, \n",
      "                   start_mid, \n",
      "                   start_max, \n",
      "                   stop_min, \n",
      "                   stop_mid, \n",
      "                   stop_max, \n",
      "                   length_min, \n",
      "                   length_mid, \n",
      "                   length_max, \n",
      "                   support]\n",
      "            tbl += [row]\n",
      "    \n",
      "        except Exception as e:\n",
      "            log(e)\n",
      "            log(traceback.format_exc(10))\n",
      "                    \n",
      "    return etl.wrap(tbl)\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}